Business Intelligence Project    
Name: Zehra Ahmed, Farah Inayat, Kisa Fatima, Zuha Aqib    
Date: 18-May-2025

In [ ]:
# print when the last code was run
from datetime import datetime
datetime.now().strftime("%Y-%m-%d %H:%M:%S")

In [ ]:
# Capture start time
start_time = datetime.now()

# Imports

In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

from datetime import datetime, timedelta

### Some Functions
some preliminary functions we can use in the code

# Data Loading

In [1166]:
# Load the dataset
df = pd.read_csv('../data/flights_jantojun2020_3M.csv')

pd.set_option('display.max_columns', None)

C:\Users\DELL\AppData\Local\Temp\ipykernel_19980\696202387.py:2: DtypeWarning: Columns (36) have mixed types. Specify dtype option on import or set low_memory=False.
  df = pd.read_csv('../data/flights_jantojun2020_3M.csv')


In [1167]:
df

YEAR  QUARTER  MONTH  DAY_OF_MONTH  DAY_OF_WEEK    FL_DATE  \
0        2020        1      1             1            3   1/1/2020   
1        2020        1      1             1            3   1/1/2020   
2        2020        1      1             1            3   1/1/2020   
3        2020        1      1             1            3   1/1/2020   
4        2020        1      1             1            3   1/1/2020   
...       ...      ...    ...           ...          ...        ...   
2745842  2020        2      6            13            6  6/13/2020   
2745843  2020        2      6            13            6  6/13/2020   
2745844  2020        2      6            13            6  6/13/2020   
2745845  2020        2      6            13            6  6/13/2020   
2745846  2020        2      6            13            6  6/13/2020   

        MKT_UNIQUE_CARRIER  MKT_CARRIER_FL_NUM TAIL_NUM ORIGIN  \
0                       WN                5888   N951WN    ONT   
1                       WN                6276   N467WN    ONT   
2                       WN                4598   N7885A    ONT   
3                       WN                4761   N551WN    ONT   
4                       WN                5162   N968WN    ONT   
...                    ...                 ...      ...    ...   
2745842                 AS                1209   N413AS    SEA   
2745843                 AS                1216   N929VA    PHX   
2745844                 AS                1217   N929VA    SEA   
2745845                 AS                1223   N841VA    SNA   
2745846                 AS                1229   N413AS    HNL   

        ORIGIN_CITY_NAME ORIGIN_STATE_ABR ORIGIN_STATE_NM DEST  \
0            Ontario, CA               CA      California  SFO   
1            Ontario, CA               CA      California  SFO   
2            Ontario, CA               CA      California  SJC   
3            Ontario, CA               CA      California  SJC   
4            Ontario, CA               CA      California  SJC   
...                  ...              ...             ...  ...   
2745842      Seattle, WA               WA      Washington  HNL   
2745843      Phoenix, AZ               AZ         Arizona  SEA   
2745844      Seattle, WA               WA      Washington  PHX   
2745845    Santa Ana, CA               CA      California  SEA   
2745846     Honolulu, HI               HI          Hawaii  SEA   

            DEST_CITY_NAME DEST_STATE_ABR DEST_STATE_NM  CRS_DEP_TIME  \
0        San Francisco, CA             CA    California          1810   
1        San Francisco, CA             CA    California          1150   
2             San Jose, CA             CA    California          2020   
3             San Jose, CA             CA    California          1340   
4             San Jose, CA             CA    California           915   
...                    ...            ...           ...           ...   
2745842       Honolulu, HI             HI        Hawaii          1050   
2745843        Seattle, WA             WA    Washington          1530   
2745844        Phoenix, AZ             AZ       Arizona          1100   
2745845        Seattle, WA             WA    Washington          1510   
2745846        Seattle, WA             WA    Washington          1520   

         DEP_TIME  DEP_DELAY  DEP_DELAY_NEW  DEP_DEL15  DEP_DELAY_GROUP  \
0          1851.0       41.0           41.0        1.0              2.0   
1          1146.0       -4.0            0.0        0.0             -1.0   
2          2016.0       -4.0            0.0        0.0             -1.0   
3          1350.0       10.0           10.0        0.0              0.0   
4           916.0        1.0            1.0        0.0              0.0   
...           ...        ...            ...        ...              ...   
2745842    1048.0       -2.0            0.0        0.0             -1.0   
2745843    1517.0      -13.0            0.0        0.0             -1.0   
2745844    1058.0 

In [1168]:
df.shape

(2745847, 47)

In [1169]:
# Function to get current date and time as a string
def get_current_datetime():
    return datetime.now().strftime("%Y-%m-%d %H:%M:%S")

# Data Preparation

## Data Statistics

In [1170]:
# Check for missing values
print("Missing values per column:")
df.isnull().sum()

Missing values per column:


YEAR                         0
QUARTER                      0
MONTH                        0
DAY_OF_MONTH                 0
DAY_OF_WEEK                  0
FL_DATE                      0
MKT_UNIQUE_CARRIER           0
MKT_CARRIER_FL_NUM           0
TAIL_NUM                162927
ORIGIN                       0
ORIGIN_CITY_NAME             0
ORIGIN_STATE_ABR             0
ORIGIN_STATE_NM              0
DEST                         0
DEST_CITY_NAME               0
DEST_STATE_ABR               0
DEST_STATE_NM                0
CRS_DEP_TIME                 0
DEP_TIME                281937
DEP_DELAY               281992
DEP_DELAY_NEW           281992
DEP_DEL15               281992
DEP_DELAY_GROUP         281992
DEP_TIME_BLK                 0
TAXI_OUT                282493
WHEELS_OFF              282493
WHEELS_ON               283554
TAXI_IN                 283554
CRS_ARR_TIME                 0
ARR_TIME                283554
ARR_DELAY               287334
ARR_DELAY_NEW           287334
ARR_DEL1

In [1171]:
# Display data types
print("\nData types:")
df.dtypes


Data types:


YEAR                     int64
QUARTER                  int64
MONTH                    int64
DAY_OF_MONTH             int64
DAY_OF_WEEK              int64
FL_DATE                 object
MKT_UNIQUE_CARRIER      object
MKT_CARRIER_FL_NUM       int64
TAIL_NUM                object
ORIGIN                  object
ORIGIN_CITY_NAME        object
ORIGIN_STATE_ABR        object
ORIGIN_STATE_NM         object
DEST                    object
DEST_CITY_NAME          object
DEST_STATE_ABR          object
DEST_STATE_NM           object
CRS_DEP_TIME             int64
DEP_TIME               float64
DEP_DELAY              float64
DEP_DELAY_NEW          float64
DEP_DEL15              float64
DEP_DELAY_GROUP        float64
DEP_TIME_BLK            object
TAXI_OUT               float64
WHEELS_OFF             float64
WHEELS_ON              float64
TAXI_IN                float64
CRS_ARR_TIME             int64
ARR_TIME               float64
ARR_DELAY              float64
ARR_DELAY_NEW          float64
ARR_DEL1

In [1172]:
# Get summary statistics
print("\nSummary statistics:")
df.describe(include='all')


Summary statistics:


YEAR       QUARTER         MONTH  DAY_OF_MONTH   DAY_OF_WEEK  \
count   2745847.0  2.745847e+06  2.745847e+06  2.745847e+06  2.745847e+06   
unique        NaN           NaN           NaN           NaN           NaN   
top           NaN           NaN           NaN           NaN           NaN   
freq          NaN           NaN           NaN           NaN           NaN   
mean       2020.0  1.277114e+00  2.811948e+00  1.548115e+01  3.963960e+00   
std           0.0  4.475736e-01  1.527229e+00  8.799480e+00  1.993749e+00   
min        2020.0  1.000000e+00  1.000000e+00  1.000000e+00  1.000000e+00   
25%        2020.0  1.000000e+00  2.000000e+00  8.000000e+00  2.000000e+00   
50%        2020.0  1.000000e+00  3.000000e+00  1.500000e+01  4.000000e+00   
75%        2020.0  2.000000e+00  4.000000e+00  2.300000e+01  6.000000e+00   
max        2020.0  2.000000e+00  6.000000e+00  3.100000e+01  7.000000e+00   

          FL_DATE MKT_UNIQUE_CARRIER  MKT_CARRIER_FL_NUM TAIL_NUM   ORIGIN  \
count     2745847            2745847        2.745847e+06  2582920  2745847   
unique        182                 10                 NaN     6079      375   
top     3/20/2020                 AA                 NaN   N494HA      ATL   
freq        23956             735559                 NaN     1106   128166   
mean          NaN                NaN        2.714583e+03      NaN      NaN   
std           NaN                NaN        1.829519e+03      NaN      NaN   
min           NaN                NaN        1.000000e+00      NaN      NaN   
25%           NaN                NaN        1.108000e+03      NaN      NaN   
50%           NaN                NaN        2.344000e+03      NaN      NaN   
75%           NaN                NaN        4.305000e+03      NaN      NaN   
max           NaN                NaN        9.888000e+03      NaN      NaN   

       ORIGIN_CITY_NAME ORIGIN_STATE_ABR ORIGIN_STATE_NM     DEST  \
count           2745847          2745847         2745847  2745847   
unique              368               52              52      375   
top         Chicago, IL               TX           Texas      ATL   
freq             159609           287693          287693   128263   
mean                NaN              NaN             NaN      NaN   
std                 NaN              NaN             NaN      NaN   
min                 NaN              NaN             NaN      NaN   
25%                 NaN              NaN             NaN      NaN   
50%                 NaN              NaN             NaN      NaN   
75%                 NaN              NaN             NaN      NaN   
max                 NaN              NaN             NaN      NaN   

       DEST_CITY_NAME DEST_STATE_ABR DEST_STATE_NM  CRS_DEP_TIME  \
count         2745847        2745847       2745847  2.745847e+06   
unique            368             52            52           NaN   
top       Chicago, IL             TX         Texas           NaN   
freq           159653         287645        287645           NaN   
mean              NaN            NaN           NaN  1.325110e+03   
std               NaN            NaN           NaN  4.786911e+02   
min               NaN            NaN           NaN  1.000000e+00   
25%               NaN            NaN           NaN  9.200000e+02   
50%               NaN            NaN           NaN  1.320000e+03   
75%               NaN            NaN           NaN  1.727000e+03   
max               NaN            NaN           NaN  2.359000e+03   

            DEP_TIME     DEP_DELAY  DEP_DELAY_NEW     DEP_DEL15  \
count   2.463910e+06  2.463855e+06   2.463855e+06  2.463855e+06   
unique           NaN           NaN            NaN           NaN   
top              NaN           NaN            NaN           NaN   
freq             NaN           NaN            NaN           NaN   
mean    1.325076e+03  3.469542e+00   8.093630e+00  1.094963e-01   
std     4.859642e+02  3.927404e+01   3.806698e+01  3.122609e-01   
min     1.000000e+00 -8.000000e+01  

## Changing Data Type

### Defined Functions
here are some functions we can use for changing the datatype

In [1173]:
def to_string(df, column_name):
    """
    Convert a pandas DataFrame column to a string.
    """
    df[column_name] = df[column_name].astype('string')
    print(f"Column Name: '{column_name}', Type: {df[column_name].dtype}")
    return df  

In [1174]:
def to_date(df, column_name, format=None):
    """
    Convert a pandas DataFrame column to datetime.
    """
    df[column_name] = pd.to_datetime(df[column_name], format=format)
    print(f"Column Name: '{column_name}', Type: {df[column_name].dtype}")
    return df

In [1175]:
def to_int(df, column_name):
    """
    Convert a pandas DataFrame column to a int.
    """
    df[column_name] = df[column_name].notna().astype('int64')
    print(f"Column Name: '{column_name}', Type: {df[column_name].dtype}")
    return df  

In [1176]:
def to_float(df, column_name):
    """
    Convert a pandas DataFrame column to a float.
    """
    df[column_name] = df[column_name].astype('float64')
    print(f"Column Name: '{column_name}', Type: {df[column_name].dtype}")
    return df  

# Column by Column Analysis
so here for each of the 47 columns, we will be doing a thourough analysis on each like whether we are keeping them, dropping them, if they need to be renamed, any datatype to be changed, are there any missing values, etc. 

In [1177]:
# a function to display all statistics of a column
def show_unique_missing_dtype(df, columnName):
    print(df[columnName].value_counts())
    print("\nmissing values: ", df[columnName].isna().sum())
    print("\ndatatype: ", df[columnName].dtype)

## YEAR

In [1178]:
show_unique_missing_dtype(df, 'YEAR')

YEAR
2020    2745847
Name: count, dtype: int64

missing values:  0

datatype:  int64


so every row is just filled with 2020's so it is useless. so better to drop the column as it does not give us any insight.

In [1179]:
df = df.drop('YEAR', axis=1)

## QUARTER

In [1180]:
show_unique_missing_dtype(df, 'QUARTER')

QUARTER
1    1984933
2     760914
Name: count, dtype: int64

missing values:  0

datatype:  int64


so this is just filled with 2 values - 1 and 2. not very effective. we can also get the QUARTER using drill down strategy using `FL_DATE` in powerBI. so better to drop.

In [1181]:
df = df.drop('QUARTER', axis=1)

## MONTH

In [1182]:
show_unique_missing_dtype(df, 'MONTH')

MONTH
3    701274
1    660556
2    623103
4    331238
6    237264
5    192412
Name: count, dtype: int64

missing values:  0

datatype:  int64


so there are no missing values, and the value counts seem fine. the datatype is fine too.

## DAY_OF_MONTH

In [1183]:
show_unique_missing_dtype(df, 'DAY_OF_MONTH')

DAY_OF_MONTH
3     97419
6     97057
5     95971
2     95345
10    93798
13    93682
9     92722
1     92291
20    91680
17    91672
4     91591
12    91523
24    91504
27    91312
7     91263
23    90779
26    90024
16    89142
19    88608
14    88460
8     88375
21    88324
28    87555
22    87379
11    86975
15    86659
29    86234
25    85630
18    85266
30    67176
31    50431
Name: count, dtype: int64

missing values:  0

datatype:  int64


so we see that majority of flights take place on 3rd of each month and the least flights tak place on 31st. theres a very large decrease of flights on 30th and 31st. now the datatype is fine, there are no missing values, and they do give some meaning. so no need to remove column.

## DAY_OF_WEEK

In [1184]:
show_unique_missing_dtype(df, 'DAY_OF_WEEK')

DAY_OF_WEEK
5    414777
4    414462
1    409181
7    394434
3    393629
2    378649
6    340715
Name: count, dtype: int64

missing values:  0

datatype:  int64


so majority of flights flew at 5:Friday and least amount of flights on 6:Saturday. not a very large difference, but there is a difference. so it does give some analysis. there are no missing values. but this is not very understandable, especially the 1-7. so lets map it to actual days, but only the first three letters to save space.

In [1185]:
# define the mapping
num_to_day = {
    1: "Mon",
    2: "Tue",
    3: "Wed",
    4: "Thu",
    5: "Fri",
    6: "Sat",
    7: "Sun"
}

# replace in your DataFrame (in-place)
df['day_of_week_c'] = df['DAY_OF_WEEK'].map(num_to_day)

In [1186]:
# Get list of current columns
cols = list(df.columns)

# Move 'day_of_week_c' to be right after 'DAY_OF_WEEK'
day_index = cols.index('DAY_OF_WEEK')
cols.insert(day_index + 1, cols.pop(cols.index('day_of_week_c')))

# Reorder DataFrame
df = df[cols]

change the datatype to string and recheck the unique values to verify that the mapping is done correctly

In [1187]:
df = to_string(df, 'day_of_week_c')

Column Name: 'day_of_week_c', Type: string


In [1188]:
show_unique_missing_dtype(df, 'day_of_week_c')

day_of_week_c
Fri    414777
Thu    414462
Mon    409181
Sun    394434
Wed    393629
Tue    378649
Sat    340715
Name: count, dtype: Int64

missing values:  0

datatype:  string


## FL_DATE

In [1189]:
show_unique_missing_dtype(df, 'FL_DATE')

FL_DATE
3/20/2020    23956
3/19/2020    23936
3/13/2020    23863
3/16/2020    23832
3/12/2020    23787
             ...  
5/16/2020     5371
5/9/2020      5347
5/5/2020      5288
5/19/2020     5251
5/12/2020     5232
Name: count, Length: 182, dtype: int64

missing values:  0

datatype:  object


so we see that the majority flights were on 20th March 2020. now convert the data to datatype `DATE`. no missing values. 

In [1190]:
# convert FL_DATE from '3/20/2020' strings to actual dates
df = to_date(df, 'FL_DATE', format='%m/%d/%Y')

Column Name: 'FL_DATE', Type: datetime64[ns]


lets rename the column to make it more clear, to `flight_date`

In [1191]:
df.rename(columns={
    'FL_DATE': 'FLIGHT_DATE'
}, inplace=True)

## MKT_UNIQUE_CARRIER

In [1192]:
show_unique_missing_dtype(df, 'MKT_UNIQUE_CARRIER')

MKT_UNIQUE_CARRIER
AA    735559
DL    551432
WN    531577
UA    499418
AS    145080
B6     88908
NK     65918
G4     53160
F9     47134
HA     27661
Name: count, dtype: int64

missing values:  0

datatype:  object


so we see that the `AA` carrier is most used whereas the `HA` carrier is least. no missing values. does give meaningful information. lets convert its `dtype` to `string`.

In [1193]:
df = to_string(df, 'MKT_UNIQUE_CARRIER')

Column Name: 'MKT_UNIQUE_CARRIER', Type: string


lets make the name more meaningful - so this column is basically the airline that MARKETS you the ticket. the airline that flies you is the OPERATING CARRIER. so lets rename as `marketing_carrier_code`    
- Marketing Carrier (MKT_UNIQUE_CARRIER): the airline that sells or markets the flight—i.e. whose flight number appears on your ticket.
- Operating Carrier (OP_CARRIER): the airline that actually operates the flight (the plane and crew).

In [1194]:
df.rename(columns={
    'MKT_UNIQUE_CARRIER': 'MARKETING_CARRIER_CODE'
}, inplace=True)

## MKT_CARRIER_FL_NUM

In [1195]:
show_unique_missing_dtype(df, 'MKT_CARRIER_FL_NUM')

MKT_CARRIER_FL_NUM
1095    1462
64      1308
403     1296
478     1249
65      1225
        ... 
9372       1
9373       1
9370       1
9400       1
9888       1
Name: count, Length: 7003, dtype: int64

missing values:  0

datatype:  int64


so `MKT_CARRIER_FL_NUM` is the flight number. we see that flight number `1095` is flown the most. no missing values. datatype is fine. again, lets rename to something more menaingful and understandable

In [1196]:
df.rename(columns={
    'MKT_CARRIER_FL_NUM': 'MARKETING_FLIGHT_NUMBER'
}, inplace=True)

## TAIL_NUM

In [1197]:
show_unique_missing_dtype(df, 'TAIL_NUM')

TAIL_NUM
N494HA    1106
N484HA    1102
N804HC    1051
N479HA    1032
N475HA    1024
          ... 
N843MH       1
N785UA       1
N626AS       1
N227UA       1
N866DA       1
Name: count, Length: 6079, dtype: int64

missing values:  162927

datatype:  object


so `tail_num` is the Aircraft Tail Number. it usually starts with N. it has 160k missing values. first lets change its datatype to string.

In [1198]:
df = to_string(df, 'TAIL_NUM')

Column Name: 'TAIL_NUM', Type: string


now lets move to missing values. first we need to check if all those flights were cancelled

In [1199]:
# 1. Filter to only the rows where TAIL_NUM is missing
missing_tail = df[df['TAIL_NUM'].isna()]

# 2. How many of those were cancelled?
cancel_counts = missing_tail['CANCELLED'].value_counts(dropna=False)

cancel_counts

CANCELLED
1    162925
0         2
Name: count, dtype: int64

In [1200]:
# Of the missing-tail flights, how many have a non-null CANCELLATION_CODE?
missing_tail['CANCELLATION_CODE'].notna().value_counts()

CANCELLATION_CODE
True     162925
False         2
Name: count, dtype: int64

oh so only 2 rows were not cancelled but majority was. so a cancelled flight will obviously not have an aircraft. so lets fill those aircraft numbers with 'CANCELLED'

In [1201]:
# Mark all cancelled flights’ TAIL_NUM as “CANCELLED”
df.loc[
    (df['TAIL_NUM'].isna()) & (df['CANCELLED'] == 1),
    'TAIL_NUM'
] = 'CANCELLED'

done. lets just check it

In [1202]:
show_unique_missing_dtype(df, 'TAIL_NUM')

TAIL_NUM
CANCELLED    162925
N494HA         1106
N484HA         1102
N804HC         1051
N479HA         1032
              ...  
N508DN            1
N864DA            1
N562NK            1
N733AR            1
N866DA            1
Name: count, Length: 6080, dtype: Int64

missing values:  2

datatype:  string


ah so still 2 flights are left. so first lets extract them and see how they look, what other thing are missing, what they could be

In [1203]:
# get the two “non-cancelled but tail‐missing” flights
problem = df[(df['TAIL_NUM'].isna())]
problem

MONTH  DAY_OF_MONTH  DAY_OF_WEEK day_of_week_c FLIGHT_DATE  \
1353545      3            31            2           Tue  2020-03-31   
1881785      3            31            2           Tue  2020-03-31   

        MARKETING_CARRIER_CODE  MARKETING_FLIGHT_NUMBER TAIL_NUM ORIGIN  \
1353545                     G4                      541     <NA>    FAR   
1881785                     G4                     1819     <NA>    PGD   

        ORIGIN_CITY_NAME ORIGIN_STATE_ABR ORIGIN_STATE_NM DEST  \
1353545        Fargo, ND               ND    North Dakota  AZA   
1881785  Punta Gorda, FL               FL         Florida  IND   

           DEST_CITY_NAME DEST_STATE_ABR DEST_STATE_NM  CRS_DEP_TIME  \
1353545       Phoenix, AZ             AZ       Arizona          1422   
1881785  Indianapolis, IN             IN       Indiana          1721   

         DEP_TIME  DEP_DELAY  DEP_DELAY_NEW  DEP_DEL15  DEP_DELAY_GROUP  \
1353545    1422.0        0.0            0.0        0.0              0.0   
1881785    1720.0       -1.0            0.0        0.0             -1.0   

        DEP_TIME_BLK  TAXI_OUT  WHEELS_OFF  WHEELS_ON  TAXI_IN  CRS_ARR_TIME  \
1353545    1400-1459       7.0      1429.0     1621.0      6.0          1527   
1881785    1700-1759      12.0      1732.0     2030.0      5.0          1941   

         ARR_TIME  ARR_DELAY  ARR_DELAY_NEW  ARR_DEL15  ARR_DELAY_GROUP  \
1353545    1627.0        NaN            NaN        NaN              NaN   
1881785    2035.0        NaN            NaN        NaN              NaN   

        ARR_TIME_BLK  CANCELLED CANCELLATION_CODE  CRS_ELAPSED_TIME  \
1353545    1500-1559          0               NaN             185.0   
1881785    1900-1959          0               NaN             140.0   

         ACTUAL_ELAPSED_TIME  AIR_TIME  DISTANCE  DISTANCE_GROUP  \
1353545                  NaN       NaN      1220               5   
1881785                  NaN       NaN       916               4   

         CARRIER_DELAY  WEATHER_DELAY  NAS_DELAY  SECURITY_DELAY  \
1353545            NaN            NaN        NaN             NaN   
1881785            NaN            NaN        NaN             NaN   

         LATE_AIRCRAFT_DELAY  
1353545                  NaN  
1881785                  NaN

so they flew on the same day = 31st march 2020 - one at 2:22pm and one at 5:21pm. both were not cancelled, just their aircraft numbers were not noted. now they are real flights and just dropping them would loose data. so lets just keep them and fill their `tail_num` with UNKNOWN.

In [1204]:
df.loc[
  (df['TAIL_NUM'].isna()) & (df['CANCELLED'] == 0),
  'TAIL_NUM'
] = 'UNKNOWN'

and lets recheck

In [1205]:
show_unique_missing_dtype(df, 'TAIL_NUM')

TAIL_NUM
CANCELLED    162925
N494HA         1106
N484HA         1102
N804HC         1051
N479HA         1032
              ...  
N857NW            1
N733AR            1
N795UA            1
N77019            1
N864DA            1
Name: count, Length: 6081, dtype: Int64

missing values:  0

datatype:  string


lets also make its name easier to understand and interpret. so lets change its name to `aircraft_tail_number`

In [1206]:
df.rename(columns={
    'TAIL_NUM': 'AIRCRAFT_TAIL_NUMBER'
}, inplace=True)

## ORIGIN

In [1207]:
show_unique_missing_dtype(df, 'ORIGIN')

ORIGIN
ATL    128166
ORD    127858
DFW    112065
DEN    101678
CLT     98762
        ...  
PPG        26
CDB        26
MVY        17
UIN        10
BKG         9
Name: count, Length: 375, dtype: int64

missing values:  0

datatype:  object


so this is Flight Departure 3-Letter Airport Abbreviation. no missing values. we do get insights of the most flew-out of city and least flew-out city - data is meaningful. lets just change the datatype to string. 

In [1208]:
df = to_string(df, 'ORIGIN')

Column Name: 'ORIGIN', Type: string


## ORIGIN_CITY_NAME

In [1209]:
show_unique_missing_dtype(df, 'ORIGIN_CITY_NAME')

ORIGIN_CITY_NAME
Chicago, IL              159609
Atlanta, GA              128166
Dallas/Fort Worth, TX    112065
Denver, CO               101678
Charlotte, NC             98762
                          ...  
Cold Bay, AK                 26
Pago Pago, TT                26
Martha's Vineyard, MA        17
Quincy, IL                   10
Branson, MO                   9
Name: count, Length: 368, dtype: int64

missing values:  0

datatype:  object


before looking at the values, lets make the datatype as string

In [1210]:
df = to_string(df, 'ORIGIN_CITY_NAME')

Column Name: 'ORIGIN_CITY_NAME', Type: string


so thats very weird. according to the previous `origin`, we see that 
- `Atlanta, GA` corresponds to `ATL`. 
- `Dallas/Fort Worth, TX` corresponds to `DFW`. 
- `Denver, CO` for `DEN`. 
- `Charlotte, NC` for `CLT`. 
- `Cold Bay, AK` for `CDB`. 
- `Pago Pago, TT` for `PPG`. 
- `Martha's Vineyard, MA` for `MVY`. 
- `Quincy, IL` for `UIN`. 
- `Branson, MO` for `BKG`.    
 
BUT      
`Chicago, IL` doesnt correspond to any. `ORD` does not correspond to any. which is fishy - data is inconsistent.   

so lets do this. lets get each unique value in `origin` and extract their rows. then lets extract all the `value_counts ()` and see if they have more than one unique value. if yes, lets see it, if not, ignore.

In [1211]:
# # 1. How many unique ORIGIN values?
# num_origins = df['ORIGIN'].nunique()
# print(f"Number of unique ORIGIN codes: {num_origins}")

# # 2. Iterate over each unique ORIGIN value
# unique_col1 = df['ORIGIN'].unique()
# count = 0
# for col1_name in unique_col1:
#     rows = df[df['ORIGIN'] == col1_name]
#     num_uniques = rows['ORIGIN_CITY_NAME'].nunique()
#     count+=1
#     if num_uniques > 1:
#         print(f"Processing ORIGIN = {col1_name}")
#         print(rows['ORIGIN_CITY_NAME'].value_counts())
#     if count % 20 == 0:
#         print("processed", count)

so basically we saw that all the `ORIGIN` map to one city. now lets do this: lets just explore the one we were fishy about - `ORIGIN = ORD`

In [1212]:
rows = df[df['ORIGIN'] == 'ORD']
rows['ORIGIN_CITY_NAME'].value_counts()

ORIGIN_CITY_NAME
Chicago, IL    127858
Name: count, dtype: Int64

so looks like `Chicago, IL` for `ORD` is fine. lets do the opposite now - for every `origin_city_name`, lets find its corresponding `origin` and extract all the >1 unique names. 

In [1213]:
# # 1. How many unique ORIGIN_CITY_NAME values?
# num_origins = df['ORIGIN_CITY_NAME'].nunique()
# print(f"Number of unique ORIGIN_CITY_NAME codes: {num_origins}")

# # 2. Iterate over each unique ORIGIN_CITY_NAME value
# unique_col1 = df['ORIGIN_CITY_NAME'].unique()
# count = 0
# for col1_name in unique_col1:
#     rows = df[df['ORIGIN_CITY_NAME'] == col1_name]
#     num_uniques = rows['ORIGIN'].nunique()
#     count+=1
#     if num_uniques > 1:
#         print(f"Processing ORIGIN_CITY_NAME = {col1_name}")
#         print(rows['ORIGIN'].value_counts())
#     if count % 20 == 0:
#         print("----processed", count)

so heres the tea. so after a little searching on internet (and chatGPT) multiple `origin` code does not mean data inconsistency, but it means that that city had more than one airport - as `origin` is the AIRPORT ABBREVIATION. not the CITY ABBREVIATION. so this is something to note. to make this more clearer, lets rename `origin` to something more meaningful like `origin_airport_code`

In [1214]:
df.rename(columns={'ORIGIN': 'ORIGIN_AIRPORT_CODE'}, inplace=True)

lets give it a final recheck because it was altered

In [1215]:
show_unique_missing_dtype(df, 'ORIGIN_AIRPORT_CODE' )

ORIGIN_AIRPORT_CODE
ATL    128166
ORD    127858
DFW    112065
DEN    101678
CLT     98762
        ...  
PPG        26
CDB        26
MVY        17
UIN        10
BKG         9
Name: count, Length: 375, dtype: Int64

missing values:  0

datatype:  string


## ORIGIN_STATE_ABR

In [1216]:
show_unique_missing_dtype(df, 'ORIGIN_STATE_ABR')

ORIGIN_STATE_ABR
TX    287693
CA    284941
FL    227389
IL    168105
GA    137830
NC    137163
NY    120614
CO    116040
VA    103216
WA     79706
AZ     79653
PA     79450
MI     74629
NV     69339
MN     53996
TN     51261
MO     51210
NJ     47738
MA     47560
UT     46158
OH     41282
HI     40572
MD     38523
OR     36189
KY     30419
LA     29627
IN     25301
WI     24849
SC     23906
OK     15930
AL     15140
AK     13447
NE     12379
ID     12060
AR     11808
IA     11614
MT     10894
NM     10869
PR     10706
CT     10217
ND      7961
KS      6576
RI      6279
ME      6199
SD      6127
MS      5832
WY      4312
NH      4032
VT      3541
WV      3040
VI      2012
TT       513
Name: count, dtype: int64

missing values:  0

datatype:  object


so first off, lets change the datatype to string

In [1217]:
df = to_string(df, 'ORIGIN_STATE_ABR')

Column Name: 'ORIGIN_STATE_ABR', Type: string


so no missing values. this is a good thing. so this column basically tells us the Flight Departure 2-Letter State Abbreviation. So we can use this column to get the state of the flight departure. this is meaningful, we will keep it, but lets rename the column to something more meaningful like `origin_state_code`

In [1218]:
df.rename(columns={
    'ORIGIN_STATE_ABR': 'ORIGIN_STATE_CODE'
}, inplace=True)

## ORIGIN_STATE_NM

In [1219]:
show_unique_missing_dtype(df, 'ORIGIN_STATE_NM')

ORIGIN_STATE_NM
Texas                                             287693
California                                        284941
Florida                                           227389
Illinois                                          168105
Georgia                                           137830
North Carolina                                    137163
New York                                          120614
Colorado                                          116040
Virginia                                          103216
Washington                                         79706
Arizona                                            79653
Pennsylvania                                       79450
Michigan                                           74629
Nevada                                             69339
Minnesota                                          53996
Tennessee                                          51261
Missouri                                           51210
New Jersey     

so lets change its datatype to string

In [1220]:
df = to_string(df, 'ORIGIN_STATE_NM')

Column Name: 'ORIGIN_STATE_NM', Type: string


so no missing values. this is the same as the last column however this is full name. lets check does it map 1 to 1? we need to check for any inconsistencies because last time we found the differenc in airport codes and city names. 

In [1221]:
column1 = 'ORIGIN_STATE_CODE'
column2 = 'ORIGIN_STATE_NM'

# 1. How many unique column1 values?
num_origins = df[column1].nunique()
print(f"Number of unique", column1, "codes: {num_origins}")

# 2. Iterate over each unique column1 value
unique_col1 = df[column1].unique()
count = 0
for col1_name in unique_col1:
    rows = df[df[column1] == col1_name]
    num_uniques = rows[column2].nunique()
    count+=1
    if num_uniques > 1:
        print(f"Processing ", column1, " = {col1_name}")
        print(rows[column2].value_counts())
    if count % 20 == 0:
        print("processed", count)

Number of unique ORIGIN_STATE_CODE codes: {num_origins}
processed 20
processed 40


In [1222]:
column1 = 'ORIGIN_STATE_NM'
column2 = 'ORIGIN_STATE_CODE'

# 1. How many unique column1 values?
num_origins = df[column1].nunique()
print(f"Number of unique", column1, "codes: {num_origins}")

# 2. Iterate over each unique column1 value
unique_col1 = df[column1].unique()
count = 0
for col1_name in unique_col1:
    rows = df[df[column1] == col1_name]
    num_uniques = rows[column2].nunique()
    count+=1
    if num_uniques > 1:
        print(f"Processing ", column1, " = {col1_name}")
        print(rows[column2].value_counts())
    if count % 20 == 0:
        print("processed", count)

Number of unique ORIGIN_STATE_NM codes: {num_origins}
processed 20
processed 40


so looks like all the data is consistent and maps to the same one. This is a good thing. lets just rename the column to something more meaningful

In [1223]:
df.rename(columns={
    'ORIGIN_STATE_NM': 'ORIGIN_STATE_NAME'
}, inplace=True)

## DEST

In [1224]:
show_unique_missing_dtype(df, 'DEST')

DEST
ATL    128263
ORD    127904
DFW    112018
DEN    101737
CLT     98727
        ...  
CDB        26
PPG        26
MVY        17
BKG         9
UIN         9
Name: count, Length: 375, dtype: int64

missing values:  0

datatype:  object


so this is again AIRPORT CODES and not CITY CODES. lets rename this to `DEST_AIRPORT_CODE` 

In [1225]:
df.rename(columns={
    'DEST': 'DEST_AIRPORT_CODE'
}, inplace=True)

and now lets change the datatype to string

In [1226]:
df = to_string(df, 'DEST_AIRPORT_CODE')

Column Name: 'DEST_AIRPORT_CODE', Type: string


## DEST_CITY_NAME

In [1227]:
show_unique_missing_dtype(df, 'DEST_CITY_NAME')

DEST_CITY_NAME
Chicago, IL              159653
Atlanta, GA              128263
Dallas/Fort Worth, TX    112018
Denver, CO               101737
Charlotte, NC             98727
                          ...  
Pago Pago, TT                26
Cold Bay, AK                 26
Martha's Vineyard, MA        17
Quincy, IL                    9
Branson, MO                   9
Name: count, Length: 368, dtype: int64

missing values:  0

datatype:  object


so no missing values. the data is same as we discussed for origin, this is destination city name. so name is menaingful and changing is not needed. so lets change datatype to string

In [1228]:
df = to_string(df, 'DEST_CITY_NAME')

Column Name: 'DEST_CITY_NAME', Type: string


## DEST_STATE_ABR

In [1229]:
show_unique_missing_dtype(df, 'DEST_STATE_ABR')

DEST_STATE_ABR
TX    287645
CA    285003
FL    227236
IL    168148
GA    137917
NC    137123
NY    120619
CO    116071
VA    103214
WA     79713
AZ     79629
PA     79440
MI     74676
NV     69322
MN     54017
TN     51258
MO     51207
NJ     47806
MA     47589
UT     46217
OH     41279
HI     40550
MD     38502
OR     36184
KY     30420
LA     29637
IN     25281
WI     24836
SC     23906
OK     15929
AL     15147
AK     13450
NE     12381
ID     11985
AR     11824
IA     11614
MT     10879
NM     10869
PR     10680
CT     10213
ND      7957
KS      6584
RI      6280
ME      6201
SD      6134
MS      5830
WY      4307
NH      4032
VT      3543
WV      3038
VI      2013
TT       512
Name: count, dtype: int64

missing values:  0

datatype:  object


so this is fine. no missing values. lets change the datatype to string

In [1230]:
df = to_string(df, 'DEST_STATE_ABR')

Column Name: 'DEST_STATE_ABR', Type: string


lets rename the column in the same way for origin so that its consistent

In [1231]:
df.rename(columns={
    'DEST_STATE_ABR': 'DEST_STATE_CODE'
}, inplace=True)

## DEST_STATE_NM 

In [1232]:
show_unique_missing_dtype(df, 'DEST_STATE_NM')

DEST_STATE_NM
Texas                                             287645
California                                        285003
Florida                                           227236
Illinois                                          168148
Georgia                                           137917
North Carolina                                    137123
New York                                          120619
Colorado                                          116071
Virginia                                          103214
Washington                                         79713
Arizona                                            79629
Pennsylvania                                       79440
Michigan                                           74676
Nevada                                             69322
Minnesota                                          54017
Tennessee                                          51258
Missouri                                           51207
New Jersey       

so no missing values. lets change datatype and rename the same way as origin

In [1233]:
df = to_string(df, 'DEST_STATE_NM')

Column Name: 'DEST_STATE_NM', Type: string


In [1234]:
df.rename(columns={
    'DEST_STATE_NM': 'DEST_STATE_NAME'
}, inplace=True)

## CRS_DEP_TIME

In [1235]:
show_unique_missing_dtype(df, 'CRS_DEP_TIME')

CRS_DEP_TIME
600    54560
700    48071
800    29759
900    19845
830    19290
       ...  
452        1
332        1
146        1
137        1
357        1
Name: count, Length: 1310, dtype: int64

missing values:  0

datatype:  int64


so no missing values. this is the SCHEDULED DEP TIME. majority flights happen at 6am and at whole numbers like 6am, 7am, 8am etc. but this is not very interpretable, we cannot make any understanding from this. but what is more interpretable is that we get the HOURS of departure. i.e. 4:52 means 4. 1:46 means 1. etc. i want to be able to interpret what is the most hour of departure flights? same for arrival.

In [1236]:
def extract_hour(df: pd.DataFrame, time_col: str, hour_col: str) -> pd.DataFrame:
    """
    Given a DataFrame and a column like 'CRS_DEP_TIME' or 'CRS_ARR_TIME' 
    in HMM or HHMM integer form, creates a new column with just the hour.

    Examples:
      3    → 0   (00:03)
      614  → 6   (06:14)
      2300 → 23  (23:00)
    """
    # ensure numeric, coerce errors to NaN
    times = pd.to_numeric(df[time_col], errors='coerce').fillna(0).astype(int)
    
    # integer-divide by 100 to get the hour bucket
    df[hour_col] = times // 100
    
    # Get list of current columns
    cols = list(df.columns)

    # Move hour_col to be right after time_col
    day_index = cols.index(time_col)
    cols.insert(day_index + 1, cols.pop(cols.index(hour_col)))
    
    # Reorder DataFrame
    df = df[cols]
    
    return df

so lets make a new column which extracts the times

In [1237]:
df = extract_hour(df, 'CRS_DEP_TIME', 'CRS_DEP_HOUR')

lets review its statistics and recheck

In [1238]:
show_unique_missing_dtype(df, 'CRS_DEP_HOUR')

CRS_DEP_HOUR
7     188892
17    185129
6     184951
8     183788
11    176881
9     173659
12    172998
15    170027
10    167429
13    164308
14    163822
16    163611
18    162903
19    145134
20    128942
21     73945
5      58407
22     57798
23     15267
0       4235
1       2260
2        671
3        509
4        281
Name: count, dtype: int64

missing values:  0

datatype:  int32


after this, we can even make a DAY/NIGHT column in `DAX`. right now its not necessary. now lets rename both column names to make it more understandable.

In [1239]:
df.rename(columns={
    'CRS_DEP_TIME': 'SCHEDULED_DEP_TIME',
    'CRS_DEP_HOUR': 'SCHEDULED_DEP_HOUR'
}, inplace=True)

## DEP_TIME

In [1240]:
show_unique_missing_dtype(df, 'DEP_TIME')

DEP_TIME
555.0    6360
655.0    5925
554.0    5466
556.0    5420
656.0    5411
         ... 
430.0       3
411.0       3
313.0       3
326.0       3
407.0       2
Name: count, Length: 1440, dtype: int64

missing values:  281937

datatype:  float64


so there are missing values in the data. lets deal with these in a bit. first lets deal with datatype - why is it float ? lets check if all values are whole floats so that we can easily convert to int?

In [1241]:
def check_if_int(df, col):
    # 1. Check for non-null values that aren’t whole numbers
    #    This will be False if any x has a fractional part.
    is_every_int = (df[col]
                    .dropna()              # ignore NaNs
                    .mod(1)                # x % 1
                    .eq(0)                 # check == 0
                    .all()                 # all True?
                )

    print(f"All non-missing ", col, " values are integers? {is_every_int}")
    return is_every_int

is_every_int = check_if_int(df, 'DEP_TIME')

All non-missing  DEP_TIME  values are integers? {is_every_int}


ah so they are whole decimals. lets convert datatype to integer

In [1242]:
def convert_int(is_every_int, df, col):
    # 2. If True, convert to an integer dtype.
    if is_every_int:
        if df[col].isna().sum() > 0:
            # If you still have NaNs, use the pandas nullable Int64 dtype:
            df[col] = df[col].astype('Int64')
        else:
            # Otherwise, for no-missing data, you can do:
            df[col] = df[col].astype('Int64')
        print("converted successfully")
    else:
        print("Warning: some ", col, " values have fractional parts—inspect those before casting.")
    return df

df = convert_int(is_every_int, df, 'DEP_TIME')

converted successfully


In [1243]:
def convert_to_int(df, column):
    int = check_if_int(df, column)
    df = convert_int(int, df, column)
    return df

now lets revisit its statistics

In [1244]:
show_unique_missing_dtype(df, 'DEP_TIME')

DEP_TIME
555    6360
655    5925
554    5466
556    5420
656    5411
       ... 
430       3
411       3
313       3
326       3
407       2
Name: count, Length: 1440, dtype: Int64

missing values:  281937

datatype:  Int64


great, so now lets deal with missing values. lets extract all the missing rows and check - are they all cancelled?

In [1245]:
missing_rows = df[df['DEP_TIME'].isna()]
# 2. How many of those were cancelled?
cancel_counts = missing_rows['CANCELLED'].value_counts(dropna=False)

cancel_counts

CANCELLED
1    281937
Name: count, dtype: int64

In [1246]:
def check_if_cancelled(df, col):
    missing_rows = df[df[col].isna()]
    # 2. How many of those were cancelled?
    print(missing_rows['CANCELLED'].value_counts(dropna=False))
    return missing_rows

ah so all were cancelled. lets fill with -1 to show its not departed

In [1247]:
missing_idx = missing_rows.index
df.loc[missing_idx, 'DEP_TIME'] = -1

In [1248]:
show_unique_missing_dtype(df, 'DEP_TIME')

DEP_TIME
-1     281937
555      6360
655      5925
554      5466
556      5420
        ...  
430         3
411         3
313         3
326         3
407         2
Name: count, Length: 1441, dtype: Int64

missing values:  0

datatype:  Int64


lets do the same for this, extract the hour

In [1249]:
df = extract_hour(df, 'DEP_TIME', 'DEP_HOUR')

In [1250]:
show_unique_missing_dtype(df, 'DEP_HOUR')

DEP_HOUR
-1     281937
 17    163727
 11    157351
 6     155794
 12    155225
 9     152406
 8     152213
 7     151386
 15    151011
 16    147706
 14    147460
 10    146673
 13    146259
 18    141828
 19    131332
 20    116144
 5      90432
 21     70179
 22     52109
 23     16845
 4       7823
 0       5942
 1       2668
 2        754
 3        540
 24       103
Name: count, dtype: int64

missing values:  0

datatype:  int32


okay so there is a problem - how can there be the hour "24"? looks like some data inconsistency is present which is causing it to be 24. lets extract and inspect those rows first

In [1251]:
# ensure the column is numeric
df['DEP_TIME_CLEAN'] = pd.to_numeric(df['DEP_TIME'], errors='coerce')

# filter rows with values > 2400
bad_times = df[df['DEP_TIME_CLEAN'] > 2400]
bad_times

Empty DataFrame
Columns: [MONTH, DAY_OF_MONTH, DAY_OF_WEEK, day_of_week_c, FLIGHT_DATE, MARKETING_CARRIER_CODE, MARKETING_FLIGHT_NUMBER, AIRCRAFT_TAIL_NUMBER, ORIGIN_AIRPORT_CODE, ORIGIN_CITY_NAME, ORIGIN_STATE_CODE, ORIGIN_STATE_NAME, DEST_AIRPORT_CODE, DEST_CITY_NAME, DEST_STATE_CODE, DEST_STATE_NAME, SCHEDULED_DEP_TIME, SCHEDULED_DEP_HOUR, DEP_TIME, DEP_HOUR, DEP_DELAY, DEP_DELAY_NEW, DEP_DEL15, DEP_DELAY_GROUP, DEP_TIME_BLK, TAXI_OUT, WHEELS_OFF, WHEELS_ON, TAXI_IN, CRS_ARR_TIME, ARR_TIME, ARR_DELAY, ARR_DELAY_NEW, ARR_DEL15, ARR_DELAY_GROUP, ARR_TIME_BLK, CANCELLED, CANCELLATION_CODE, CRS_ELAPSED_TIME, ACTUAL_ELAPSED_TIME, AIR_TIME, DISTANCE, DISTANCE_GROUP, CARRIER_DELAY, WEATHER_DELAY, NAS_DELAY, SECURITY_DELAY, LATE_AIRCRAFT_DELAY, DEP_TIME_CLEAN]
Index: []

none. lets extract all those with equal to 2400

In [1252]:
# ensure the column is numeric
df['DEP_TIME_CLEAN'] = pd.to_numeric(df['DEP_TIME'], errors='coerce')

# filter rows with values = 2400
bad_times = df[df['DEP_TIME_CLEAN'] == 2400]
bad_times

MONTH  DAY_OF_MONTH  DAY_OF_WEEK day_of_week_c FLIGHT_DATE  \
71387        1            28            2           Tue  2020-01-28   
77444        1             9            4           Thu  2020-01-09   
77461        1            26            7           Sun  2020-01-26   
81292        1             7            2           Tue  2020-01-07   
84223        1            17            5           Fri  2020-01-17   
...        ...           ...          ...           ...         ...   
1828856      3             2            1           Mon  2020-03-02   
1881502      3            16            1           Mon  2020-03-16   
2183266      4             4            6           Sat  2020-04-04   
2270267      4             1            3           Wed  2020-04-01   
2507831      5            24            7           Sun  2020-05-24   

        MARKETING_CARRIER_CODE  MARKETING_FLIGHT_NUMBER AIRCRAFT_TAIL_NUMBER  \
71387                       F9                      558               N202FR   
77444                       NK                     1839               N646NK   
77461                       NK                     1839               N647NK   
81292                       DL                     1195               N846DN   
84223                       DL                      386               N312DN   
...                        ...                      ...                  ...   
1828856                     AA                     1151               N816NN   
1881502                     G4                      269                224NV   
2183266                     AA                      340               N987AM   
2270267                     AA                     1004               N934AA   
2507831                     G4                        5                304NV   

        ORIGIN_AIRPORT_CODE   ORIGIN_CITY_NAME ORIGIN_STATE_CODE  \
71387                   DEN         Denver, CO                CO   
77444                   LAS      Las Vegas, NV                NV   
77461                   LAS      Las Vegas, NV                NV   
81292                   LAX    Los Angeles, CA                CA   
84223                   ATL        Atlanta, GA                GA   
...                     ...                ...               ...   
1828856                 MIA          Miami, FL                FL   
1881502                 SFB        Sanford, FL                FL   
2183266                 LAX    Los Angeles, CA                CA   
2270267                 SFO  San Francisco, CA                CA   
2507831                 ELP        El Paso, TX                TX   

        ORIGIN_STATE_NAME DEST_AIRPORT_CODE                  DEST_CITY_NAME  \
71387            Colorado               RDU              Raleigh/Durham, NC   
77444              Nevada               BNA                   Nashville, TN   
77461              Nevada               BNA                   Nashville, TN   
81292          California               ATL                     Atlanta, GA   
84223             Georgia               TPA                       Tampa, FL   
...                   ...               ...                             ...   
1828856           Florida               DFW           Dallas/Fort Worth, TX   
1881502           Florida               ABE  Allentown/Bethlehem/Easton, PA   
2183266        California               MIA                       Miami, FL   
2270267        California               DFW           Dallas/Fort Worth, TX   
2507831             Texas               LAS                   Las Vegas, NV   

        DEST_STATE_CODE DEST_STATE_NAME  SCHEDULED_DEP_TIME  \
71387                NC  North Carolina                   2   
77444                TN       Tennessee                2359   
77461                TN       Tennessee                2359   
81292                GA         Georgia                2355   
84223                FL         Florida                2330   
...                 ...             ...            

ah so it looks like all 2400 values mean raat 12 baje - i.e. 00:00. lets map all those values to 00:00 first

In [1253]:
# ensure numeric dtype
df['DEP_TIME'] = pd.to_numeric(df['DEP_TIME'], errors='coerce').astype('Int64')

# replace exactly 2400 → 0
df.loc[df['DEP_TIME'] == 2400, 'DEP_TIME'] = 0

In [1254]:
# RECHECK

# ensure the column is numeric
df['DEP_TIME_CLEAN'] = pd.to_numeric(df['DEP_TIME'], errors='coerce')

# filter rows with values = 2400
bad_times = df[df['DEP_TIME_CLEAN'] == 2400]
bad_times

Empty DataFrame
Columns: [MONTH, DAY_OF_MONTH, DAY_OF_WEEK, day_of_week_c, FLIGHT_DATE, MARKETING_CARRIER_CODE, MARKETING_FLIGHT_NUMBER, AIRCRAFT_TAIL_NUMBER, ORIGIN_AIRPORT_CODE, ORIGIN_CITY_NAME, ORIGIN_STATE_CODE, ORIGIN_STATE_NAME, DEST_AIRPORT_CODE, DEST_CITY_NAME, DEST_STATE_CODE, DEST_STATE_NAME, SCHEDULED_DEP_TIME, SCHEDULED_DEP_HOUR, DEP_TIME, DEP_HOUR, DEP_DELAY, DEP_DELAY_NEW, DEP_DEL15, DEP_DELAY_GROUP, DEP_TIME_BLK, TAXI_OUT, WHEELS_OFF, WHEELS_ON, TAXI_IN, CRS_ARR_TIME, ARR_TIME, ARR_DELAY, ARR_DELAY_NEW, ARR_DEL15, ARR_DELAY_GROUP, ARR_TIME_BLK, CANCELLED, CANCELLATION_CODE, CRS_ELAPSED_TIME, ACTUAL_ELAPSED_TIME, AIR_TIME, DISTANCE, DISTANCE_GROUP, CARRIER_DELAY, WEATHER_DELAY, NAS_DELAY, SECURITY_DELAY, LATE_AIRCRAFT_DELAY, DEP_TIME_CLEAN]
Index: []

In [1255]:
# remove dep_time_clean

df = df.drop('DEP_TIME_CLEAN', axis=1)

lets re-evaluate dep_hour now

In [1256]:
df = extract_hour(df, 'DEP_TIME', 'DEP_HOUR')

In [1257]:
show_unique_missing_dtype(df, 'DEP_HOUR')

DEP_HOUR
-1     281937
 17    163727
 11    157351
 6     155794
 12    155225
 9     152406
 8     152213
 7     151386
 15    151011
 16    147706
 14    147460
 10    146673
 13    146259
 18    141828
 19    131332
 20    116144
 5      90432
 21     70179
 22     52109
 23     16845
 4       7823
 0       6045
 1       2668
 2        754
 3        540
Name: count, dtype: int64

missing values:  0

datatype:  int32


and lets rename both columns to improve readability

In [1258]:
df.rename(columns={
    'DEP_TIME': 'ACTUAL_DEP_TIME',
    'DEP_HOUR': 'ACTUAL_DEP_HOUR'
}, inplace=True)

## DEP_DELAY

In [1259]:
show_unique_missing_dtype(df, 'DEP_DELAY')

DEP_DELAY
-5.0       229122
-4.0       201527
-6.0       194063
-3.0       180863
-7.0       168413
            ...  
 1203.0         1
 850.0          1
 1136.0         1
 953.0          1
 849.0          1
Name: count, Length: 1371, dtype: int64

missing values:  281992

datatype:  float64


so lets check if its an integer or not and convert accordingly

In [1260]:
df = convert_to_int(df, 'DEP_DELAY')

All non-missing  DEP_DELAY  values are integers? {is_every_int}
converted successfully


In [1261]:
show_unique_missing_dtype(df, 'DEP_DELAY')

DEP_DELAY
-5      229122
-4      201527
-6      194063
-3      180863
-7      168413
         ...  
1203         1
850          1
1136         1
953          1
849          1
Name: count, Length: 1371, dtype: Int64

missing values:  281992

datatype:  Int64


so now lets deal with missing values by extracting them and their cancelled values

In [1262]:
missing_rows = check_if_cancelled(df, 'DEP_DELAY')

CANCELLED
1    281992
Name: count, dtype: int64


In [1263]:
missing_rows.isna().sum()

MONTH                           0
DAY_OF_MONTH                    0
DAY_OF_WEEK                     0
day_of_week_c                   0
FLIGHT_DATE                     0
MARKETING_CARRIER_CODE          0
MARKETING_FLIGHT_NUMBER         0
AIRCRAFT_TAIL_NUMBER            0
ORIGIN_AIRPORT_CODE             0
ORIGIN_CITY_NAME                0
ORIGIN_STATE_CODE               0
ORIGIN_STATE_NAME               0
DEST_AIRPORT_CODE               0
DEST_CITY_NAME                  0
DEST_STATE_CODE                 0
DEST_STATE_NAME                 0
SCHEDULED_DEP_TIME              0
SCHEDULED_DEP_HOUR              0
ACTUAL_DEP_TIME                 0
ACTUAL_DEP_HOUR                 0
DEP_DELAY                  281992
DEP_DELAY_NEW              281992
DEP_DEL15                  281992
DEP_DELAY_GROUP            281992
DEP_TIME_BLK                    0
TAXI_OUT                   281992
WHEELS_OFF                 281992
WHEELS_ON                  281992
TAXI_IN                    281992
CRS_ARR_TIME  

so what do we fill them with? if we put `DEP_DELAY` as -1, it means all our flights were 1 minute early which skews and disrupts our analysis. so we need to fill with such a value that provides us a clear distinction that all these are cancelled. so lets try and extract the min/max values of the column

In [1264]:
min_delay = df['DEP_DELAY'].min()
max_delay = df['DEP_DELAY'].max()

print("min", min_delay)
print("max", max_delay)

min -80
max 2814


okay, lets also extract all the cancelled rows and see what they are filled with

In [1265]:
cancelled_all = df[df['CANCELLED'] == 1]
show_unique_missing_dtype(cancelled_all, 'DEP_DELAY')

DEP_DELAY
-5     66
-6     54
-4     54
-7     44
-8     42
       ..
172     1
266     1
283     1
75      1
169     1
Name: count, Length: 198, dtype: Int64

missing values:  281992

datatype:  Int64


In [1266]:
cancelled_all

MONTH  DAY_OF_MONTH  DAY_OF_WEEK day_of_week_c FLIGHT_DATE  \
35           1             1            3           Wed  2020-01-01   
839          1             1            3           Wed  2020-01-01   
1266         1            20            1           Mon  2020-01-20   
1283         1            20            1           Mon  2020-01-20   
1323         1            30            4           Thu  2020-01-30   
...        ...           ...          ...           ...         ...   
2744984      6             9            2           Tue  2020-06-09   
2744985      6             9            2           Tue  2020-06-09   
2745802      6            13            6           Sat  2020-06-13   
2745803      6            13            6           Sat  2020-06-13   
2745825      6            13            6           Sat  2020-06-13   

        MARKETING_CARRIER_CODE  MARKETING_FLIGHT_NUMBER AIRCRAFT_TAIL_NUMBER  \
35                          WN                     6173               N7831B   
839                         WN                     4433               N488WN   
1266                        WN                      423               N209WN   
1283                        WN                     1519               N467WN   
1323                        WN                     1845               N741SA   
...                        ...                      ...                  ...   
2744984                     AS                      919               N612AS   
2744985                     AS                      920               N612AS   
2745802                     AS                      919               N615AS   
2745803                     AS                      920               N615AS   
2745825                     AS                     1116               N431AS   

        ORIGIN_AIRPORT_CODE ORIGIN_CITY_NAME ORIGIN_STATE_CODE  \
35                      PDX     Portland, OR                OR   
839                     SNA    Santa Ana, CA                CA   
1266                    DAL       Dallas, TX                TX   
1283                    DAL       Dallas, TX                TX   
1323                    MSY  New Orleans, LA                LA   
...                     ...              ...               ...   
2744984                 AKN  King Salmon, AK                AK   
2744985                 ANC    Anchorage, AK                AK   
2745802                 AKN  King Salmon, AK                AK   
2745803                 ANC    Anchorage, AK                AK   
2745825                 ANC    Anchorage, AK                AK   

        ORIGIN_STATE_NAME DEST_AIRPORT_CODE   DEST_CITY_NAME DEST_STATE_CODE  \
35                 Oregon               LAS    Las Vegas, NV              NV   
839            California               OAK      Oakland, CA              CA   
1266                Texas               PHX      Phoenix, AZ              AZ   
1283                Texas               SAT  San Antonio, TX              TX   
1323            Louisiana               HOU      Houston, TX              TX   
...                   ...               ...              ...             ...   
2744984            Alaska               ANC    Anchorage, AK              AK   
2744985            Alaska               AKN  King Salmon, AK              AK   
2745802            Alaska               ANC    Anchorage, AK              AK   
2745803            Alaska               AKN  King Salmon, AK              AK   
2745825            Alaska               SEA      Seattle, WA              WA   

        DEST_STATE_NAME  SCHEDULED_DEP_TIME  SCHEDULED_DEP_HOUR  \
35               Nevada                 700                   7   
839          California                 730                   7   
1266            Arizona                 600                   6   
1283              Texas                 720                   7   
1323              Texas                 605                   6   
...                 ...                

so this is giving us information that out of the 280k cancelled flights, roughly 1000 flights did take off before being cancelled - so they do have values for `dep_time` and `dep_delay`. now we dont want to loose that data. so lets fill the remaining null values with 9999 as its a unique placeholder which lets us know this flight never went. it was cancelled. we will do the same for `arr_delay`

In [1267]:
cols_to_fill = ['DEP_DELAY', 'ARR_DELAY']

In [1268]:
# Loop over each column and fill NaNs with 0 in the original df using cancelled_only's index
for col in cols_to_fill:
    missing_idx = missing_rows[missing_rows[col].isna()].index
    df.loc[missing_idx, col] = 9999

so lets make all these null columns as -1 to show that it is a cancelled flight

In [1269]:
cols_to_fill = ['DEP_DELAY_NEW', 'TAXI_OUT', 'TAXI_IN', 'WHEELS_OFF', 'WHEELS_ON', 'ARR_TIME', 'ARR_DELAY_NEW', 'CARRIER_DELAY', 'WEATHER_DELAY', 'NAS_DELAY', 'SECURITY_DELAY', 'LATE_AIRCRAFT_DELAY']

In [1270]:
# Loop over each column and fill NaNs with 0 in the original df using cancelled_only's index
for col in cols_to_fill:
    missing_idx = missing_rows[missing_rows[col].isna()].index
    df.loc[missing_idx, col] = -1

and now lets fill elapsed time and etc with 0 to indicate it was of 0 time

In [1271]:
cols_to_fill = ['ACTUAL_ELAPSED_TIME', 'AIR_TIME']

In [1272]:
# Loop over each column and fill NaNs with 0 in the original df using cancelled_only's index
for col in cols_to_fill:
    missing_idx = missing_rows[missing_rows[col].isna()].index
    df.loc[missing_idx, col] = 0

lets just give a quick recheck

In [1273]:
show_unique_missing_dtype(df, 'DEP_DELAY')

DEP_DELAY
9999    281992
-5      229122
-4      201527
-6      194063
-3      180863
         ...  
1203         1
850          1
1136         1
953          1
849          1
Name: count, Length: 1372, dtype: Int64

missing values:  0

datatype:  Int64


In [1274]:
cancelled_all = df[df['CANCELLED'] == 1]
show_unique_missing_dtype(cancelled_all, 'DEP_DELAY')

DEP_DELAY
9999    281992
-5          66
-6          54
-4          54
-7          44
         ...  
140          1
157          1
86           1
114          1
169          1
Name: count, Length: 199, dtype: Int64

missing values:  0

datatype:  Int64


In [1275]:
show_unique_missing_dtype(df, 'DEP_DELAY')

DEP_DELAY
9999    281992
-5      229122
-4      201527
-6      194063
-3      180863
         ...  
1203         1
850          1
1136         1
953          1
849          1
Name: count, Length: 1372, dtype: Int64

missing values:  0

datatype:  Int64


## DEP_DELAY_NEW

In [1276]:
show_unique_missing_dtype(df, 'DEP_DELAY_NEW')

DEP_DELAY_NEW
 0.0       1925069
-1.0        281992
 1.0         40350
 2.0         31338
 3.0         26677
            ...   
 1368.0          1
 1162.0          1
 1083.0          1
 546.0           1
 849.0           1
Name: count, Length: 1302, dtype: int64

missing values:  0

datatype:  float64


i feel as if this is just `DEP_DELAY` but all the `<0` values are put as 0. this is just the same data, repition and it can be achieved by putting a filter of greater than 0. so this column does not provide any meaningful information. we can just drop this column. 

In [1277]:
df = df.drop('DEP_DELAY_NEW', axis=1)

## DEP_DEL15

In [1278]:
show_unique_missing_dtype(df, 'DEP_DEL15')

DEP_DEL15
0.0    2194072
1.0     269783
Name: count, dtype: int64

missing values:  281992

datatype:  float64


so this is not very meaningful. lets remap it to greater than 15 and lesser than 15. but before that, lets fill the missing values. so if `dep_delay` == 9999, then its cancelled. if its < 15 then its 0, and if its > 15 then its 1. lets first fill all the missing values. 

In [1279]:
missing_rows = df[df['DEP_DEL15'].isna()]
missing_rows.shape

(281992, 47)

In [1280]:
cancelled = missing_rows[missing_rows['DEP_DELAY'] == 9999]
cancelled.shape

(281992, 47)

ah so all are cancelled, no need to fill with 0 or 1. just a direct fill of -1. 

In [1281]:
missing_idx = cancelled.index
df.loc[missing_idx, 'DEP_DEL15'] = -1

In [1282]:
show_unique_missing_dtype(df, 'DEP_DEL15')

DEP_DEL15
 0.0    2194072
-1.0     281992
 1.0     269783
Name: count, dtype: int64

missing values:  0

datatype:  float64


now lets map them

In [1283]:
cancel_map = {
    0.0: 'Less than 15',
    1.0: 'Greater than 15',
    -1.0: 'Cancelled'
}

df['DEP_DEL15'] = df['DEP_DEL15'].map(cancel_map)

In [1284]:
show_unique_missing_dtype(df, 'DEP_DEL15')

DEP_DEL15
Less than 15       2194072
Cancelled           281992
Greater than 15     269783
Name: count, dtype: int64

missing values:  0

datatype:  object


that reminds me, we need to account for all the 1000 flights which took off but were cancelled. their `dep_delay` we will not change, but we can change their map

In [1285]:
missing_rows = df[df['CANCELLED'] == 1]
missing_rows['DEP_DEL15'].value_counts()

DEP_DEL15
Cancelled          281992
Less than 15          616
Greater than 15       318
Name: count, dtype: int64

so lets change them all to cancelled

In [1286]:
missing_idx = missing_rows.index
df.loc[missing_idx, 'DEP_DEL15'] = 'Cancelled'

lets recheck

In [1287]:
missing_rows = df[df['CANCELLED'] == 1]
missing_rows['DEP_DEL15'].value_counts()

DEP_DEL15
Cancelled    282926
Name: count, dtype: int64

In [1288]:
show_unique_missing_dtype(df, 'DEP_DEL15')

DEP_DEL15
Less than 15       2193456
Cancelled           282926
Greater than 15     269465
Name: count, dtype: int64

missing values:  0

datatype:  object


lets convert datatype to string and rename the column

In [1289]:
df = to_string(df, 'DEP_DEL15')

Column Name: 'DEP_DEL15', Type: string


In [1290]:
df.rename(columns={
    'DEP_DEL15': 'DEP_DELAY_15_MIN'
}, inplace=True)

## DEP_DELAY_GROUP

In [1291]:
show_unique_missing_dtype(df, 'DEP_DELAY_GROUP')

DEP_DELAY_GROUP
-1.0     1780566
 0.0      370584
 1.0       97532
 2.0       49829
-2.0       42922
 3.0       31200
 4.0       21076
 12.0      16465
 5.0       15212
 6.0       11098
 7.0        8572
 8.0        6484
 9.0        5082
 10.0       4018
 11.0       3215
Name: count, dtype: int64

missing values:  281992

datatype:  float64


so this is 15 minute intervals. like -1 means -15 to 0, 0 means 0-15, 1 means 15-20 etc. so this isnt very readable or meaningful, better to map to strings. but first, lets fill the missing values - which by guess and proof - is cancelled flights. 

In [1292]:
missing_rows = df[df['DEP_DELAY_GROUP'].isna()]
missing_rows.shape

(281992, 47)

In [1293]:
cancelled = missing_rows[missing_rows['CANCELLED'] == 1]
cancelled.shape

(281992, 47)

yep so all are cancelled. lets fill with 9999 to show cancelled

In [1294]:
missing_idx = cancelled.index
df.loc[missing_idx, 'DEP_DELAY_GROUP'] = 9999

lets check it

In [1295]:
show_unique_missing_dtype(df, 'DEP_DELAY_GROUP')

DEP_DELAY_GROUP
-1.0       1780566
 0.0        370584
 9999.0     281992
 1.0         97532
 2.0         49829
-2.0         42922
 3.0         31200
 4.0         21076
 12.0        16465
 5.0         15212
 6.0         11098
 7.0          8572
 8.0          6484
 9.0          5082
 10.0         4018
 11.0         3215
Name: count, dtype: int64

missing values:  0

datatype:  float64


lets also analyse what all the cancelled flights are filled with

In [1296]:
cancelled_all = df[df['CANCELLED'] == 1]
cancelled_all['DEP_DELAY_GROUP'].value_counts()

DEP_DELAY_GROUP
 9999.0    281992
-1.0          470
 0.0          125
 1.0           64
 12.0          46
 2.0           41
 3.0           38
 5.0           32
 4.0           23
-2.0           21
 8.0           20
 6.0           18
 7.0           16
 11.0           8
 9.0            7
 10.0           5
Name: count, dtype: int64

ah so thats not good. lets fill all with 9999 to make it consistent

In [1297]:
missing_idx = cancelled_all.index
df.loc[missing_idx, 'DEP_DELAY_GROUP'] = 9999

In [1298]:
cancelled_all = df[df['CANCELLED'] == 1]
cancelled_all['DEP_DELAY_GROUP'].value_counts()

DEP_DELAY_GROUP
9999.0    282926
Name: count, dtype: int64

ah so we are done. lets make our map. before we do that, lets find the min/max so that we know how far and below to go

In [1299]:
min_delay = df['DEP_DELAY_GROUP'].min()
max_delay = df['DEP_DELAY_GROUP'].max()

print("min", min_delay)
print("max", max_delay)

min -2.0
max 9999.0


In [1300]:
cancel_map = {
    -2.0: '15-30 min early',
    -1.0: '0-15 min early',
    0.0: '0-15 min late', 
    1.0: '15-30 min late', 
    2.0: '30-45 min late', 
    3.0: '45-60 min late', 
    4.0: '60-75 min late', 
    5.0: '75-90 min late', 
    6.0: '90-105 min late', 
    7.0: '105-120 min late', 
    8.0: '120-135 min late', 
    9.0: '135-150 min late', 
    10.0: '150-165 min late', 
    11.0: '165-180 min late', 
    12.0: '180-195 min late',
    9999.0: 'Cancelled'
}

df['DEP_DELAY_GROUP'] = df['DEP_DELAY_GROUP'].map(cancel_map)

In [1301]:
show_unique_missing_dtype(df, 'DEP_DELAY_GROUP')

DEP_DELAY_GROUP
0-15 min early      1780096
0-15 min late        370459
Cancelled            282926
15-30 min late        97468
30-45 min late        49788
15-30 min early       42901
45-60 min late        31162
60-75 min late        21053
180-195 min late      16419
75-90 min late        15180
90-105 min late       11080
105-120 min late       8556
120-135 min late       6464
135-150 min late       5075
150-165 min late       4013
165-180 min late       3207
Name: count, dtype: int64

missing values:  0

datatype:  object


In [1302]:
df = to_string(df, 'DEP_DELAY_GROUP')

Column Name: 'DEP_DELAY_GROUP', Type: string


## DEP_TIME_BLK

In [1303]:
show_unique_missing_dtype(df, 'DEP_TIME_BLK')

DEP_TIME_BLK
0700-0759    188892
1700-1759    185129
0600-0659    184951
0800-0859    183788
1100-1159    176881
0900-0959    173659
1200-1259    172998
1500-1559    170027
1000-1059    167429
1300-1359    164308
1400-1459    163822
1600-1659    163611
1800-1859    162903
1900-1959    145134
2000-2059    128942
2100-2159     73945
0001-0559     66363
2200-2259     57798
2300-2359     15267
Name: count, dtype: int64

missing values:  0

datatype:  object


so no missing values. i think it is menaingful, but smth nice to see is that 12am - 6am is a one block, rest are hourly blocks. this is a bit inconsistent with the rest of the values. but something even more noticeable is that its the same as `SCHEDULED_DEP_HOUR`. lets see it,

In [1304]:
show_unique_missing_dtype(df, 'SCHEDULED_DEP_HOUR')

SCHEDULED_DEP_HOUR
7     188892
17    185129
6     184951
8     183788
11    176881
9     173659
12    172998
15    170027
10    167429
13    164308
14    163822
16    163611
18    162903
19    145134
20    128942
21     73945
5      58407
22     57798
23     15267
0       4235
1       2260
2        671
3        509
4        281
Name: count, dtype: int64

missing values:  0

datatype:  int32


so we dont need to convert this column, we can just drop it. but we acknowledge that this column is much more meaningful in terms of readability like 0700-0759 is much better than just 7. but then the 0700-0759 is a string datatype and makes categorical data which isnt very good.

In [1305]:
df = df.drop('DEP_TIME_BLK', axis=1)

## TAXI_OUT

In [1306]:
show_unique_missing_dtype(df, 'TAXI_OUT')

TAXI_OUT
-1.0      281992
 11.0     206870
 12.0     201378
 10.0     194639
 13.0     187982
           ...  
 160.0         1
 170.0         1
 189.0         1
 169.0         1
 180.0         1
Name: count, Length: 169, dtype: int64

missing values:  501

datatype:  float64


so -1 means cancelled flight. this is in minutes. lets handle missing values. are they cancelled flights?

In [1307]:
cancelled = check_if_cancelled(df, 'TAXI_OUT')
cancelled.isna().sum()

CANCELLED
1    501
Name: count, dtype: int64


MONTH                        0
DAY_OF_MONTH                 0
DAY_OF_WEEK                  0
day_of_week_c                0
FLIGHT_DATE                  0
MARKETING_CARRIER_CODE       0
MARKETING_FLIGHT_NUMBER      0
AIRCRAFT_TAIL_NUMBER         0
ORIGIN_AIRPORT_CODE          0
ORIGIN_CITY_NAME             0
ORIGIN_STATE_CODE            0
ORIGIN_STATE_NAME            0
DEST_AIRPORT_CODE            0
DEST_CITY_NAME               0
DEST_STATE_CODE              0
DEST_STATE_NAME              0
SCHEDULED_DEP_TIME           0
SCHEDULED_DEP_HOUR           0
ACTUAL_DEP_TIME              0
ACTUAL_DEP_HOUR              0
DEP_DELAY                    0
DEP_DELAY_15_MIN             0
DEP_DELAY_GROUP              0
TAXI_OUT                   501
WHEELS_OFF                 501
WHEELS_ON                  501
TAXI_IN                    501
CRS_ARR_TIME                 0
ARR_TIME                   501
ARR_DELAY                  501
ARR_DELAY_NEW              501
ARR_DEL15                  501
ARR_DELA

so lets fill with -1, as all times are cancelled and did not occur

In [1308]:
cols_to_fill = ['TAXI_OUT', 'TAXI_IN', 'WHEELS_OFF', 'WHEELS_ON', 'ARR_TIME', 'ARR_DELAY_NEW', 'CARRIER_DELAY', 'WEATHER_DELAY', 'NAS_DELAY', 'SECURITY_DELAY', 'LATE_AIRCRAFT_DELAY']

In [1309]:
# Loop over each column and fill NaNs with 0 in the original df using cancelled_only's index
for col in cols_to_fill:
    missing_idx = cancelled[cancelled[col].isna()].index
    df.loc[missing_idx, col] = -1

and lets fill with 0 for elapsed_times

In [1310]:
cols_to_fill = ['ACTUAL_ELAPSED_TIME', 'AIR_TIME']

In [1311]:
# Loop over each column and fill NaNs with 0 in the original df using cancelled_only's index
for col in cols_to_fill:
    missing_idx = cancelled[cancelled[col].isna()].index
    df.loc[missing_idx, col] = 0

lets recheck

In [1312]:
show_unique_missing_dtype(df, 'TAXI_OUT')

TAXI_OUT
-1.0      282493
 11.0     206870
 12.0     201378
 10.0     194639
 13.0     187982
           ...  
 160.0         1
 170.0         1
 189.0         1
 169.0         1
 180.0         1
Name: count, Length: 169, dtype: int64

missing values:  0

datatype:  float64


In [1313]:
cancelled_all = df[df['CANCELLED'] == 1]
cancelled_all['TAXI_OUT'].value_counts()

TAXI_OUT
-1.0      282493
 12.0         33
 17.0         31
 15.0         29
 14.0         26
 13.0         26
 18.0         22
 11.0         21
 10.0         21
 21.0         21
 16.0         19
 24.0         17
 9.0          17
 20.0         11
 22.0         11
 19.0         10
 30.0         10
 23.0          9
 7.0           9
 8.0           7
 25.0          7
 31.0          7
 37.0          6
 26.0          6
 29.0          5
 39.0          5
 38.0          4
 27.0          4
 33.0          4
 35.0          3
 36.0          3
 6.0           3
 28.0          3
 40.0          2
 68.0          2
 45.0          2
 41.0          2
 49.0          2
 32.0          2
 53.0          1
 46.0          1
 43.0          1
 100.0         1
 34.0          1
 79.0          1
 81.0          1
 2.0           1
 48.0          1
 60.0          1
 5.0           1
Name: count, dtype: int64

In [1314]:
cancelled_all

MONTH  DAY_OF_MONTH  DAY_OF_WEEK day_of_week_c FLIGHT_DATE  \
35           1             1            3           Wed  2020-01-01   
839          1             1            3           Wed  2020-01-01   
1266         1            20            1           Mon  2020-01-20   
1283         1            20            1           Mon  2020-01-20   
1323         1            30            4           Thu  2020-01-30   
...        ...           ...          ...           ...         ...   
2744984      6             9            2           Tue  2020-06-09   
2744985      6             9            2           Tue  2020-06-09   
2745802      6            13            6           Sat  2020-06-13   
2745803      6            13            6           Sat  2020-06-13   
2745825      6            13            6           Sat  2020-06-13   

        MARKETING_CARRIER_CODE  MARKETING_FLIGHT_NUMBER AIRCRAFT_TAIL_NUMBER  \
35                          WN                     6173               N7831B   
839                         WN                     4433               N488WN   
1266                        WN                      423               N209WN   
1283                        WN                     1519               N467WN   
1323                        WN                     1845               N741SA   
...                        ...                      ...                  ...   
2744984                     AS                      919               N612AS   
2744985                     AS                      920               N612AS   
2745802                     AS                      919               N615AS   
2745803                     AS                      920               N615AS   
2745825                     AS                     1116               N431AS   

        ORIGIN_AIRPORT_CODE ORIGIN_CITY_NAME ORIGIN_STATE_CODE  \
35                      PDX     Portland, OR                OR   
839                     SNA    Santa Ana, CA                CA   
1266                    DAL       Dallas, TX                TX   
1283                    DAL       Dallas, TX                TX   
1323                    MSY  New Orleans, LA                LA   
...                     ...              ...               ...   
2744984                 AKN  King Salmon, AK                AK   
2744985                 ANC    Anchorage, AK                AK   
2745802                 AKN  King Salmon, AK                AK   
2745803                 ANC    Anchorage, AK                AK   
2745825                 ANC    Anchorage, AK                AK   

        ORIGIN_STATE_NAME DEST_AIRPORT_CODE   DEST_CITY_NAME DEST_STATE_CODE  \
35                 Oregon               LAS    Las Vegas, NV              NV   
839            California               OAK      Oakland, CA              CA   
1266                Texas               PHX      Phoenix, AZ              AZ   
1283                Texas               SAT  San Antonio, TX              TX   
1323            Louisiana               HOU      Houston, TX              TX   
...                   ...               ...              ...             ...   
2744984            Alaska               ANC    Anchorage, AK              AK   
2744985            Alaska               AKN  King Salmon, AK              AK   
2745802            Alaska               ANC    Anchorage, AK              AK   
2745803            Alaska               AKN  King Salmon, AK              AK   
2745825            Alaska               SEA      Seattle, WA              WA   

        DEST_STATE_NAME  SCHEDULED_DEP_TIME  SCHEDULED_DEP_HOUR  \
35               Nevada                 700                   7   
839          California                 730                   7   
1266            Arizona                 600                   6   
1283              Texas                 720                   7   
1323              Texas                 605                   6   
...                 ...                

lets check if we can convert to integer

In [1315]:
df = convert_to_int(df, 'TAXI_OUT')

All non-missing  TAXI_OUT  values are integers? {is_every_int}
converted successfully


In [1316]:
show_unique_missing_dtype(df, 'TAXI_OUT')

TAXI_OUT
-1     282493
11     206870
12     201378
10     194639
13     187982
        ...  
160         1
170         1
189         1
169         1
180         1
Name: count, Length: 169, dtype: Int64

missing values:  0

datatype:  Int64


rename to something more menaingful like `ORIGIN_TAXI_TIME`

In [1317]:
df.rename(columns={'TAXI_OUT': 'ORIGIN_TAXI_TIME'}, inplace=True)

## WHEELS_OFF

In [1318]:
show_unique_missing_dtype(df, 'WHEELS_OFF')

WHEELS_OFF
-1.0      282493
 707.0      4090
 708.0      4046
 706.0      4010
 709.0      3960
           ...  
 442.0         4
 440.0         3
 447.0         3
 336.0         3
 347.0         3
Name: count, Length: 1440, dtype: int64

missing values:  0

datatype:  float64


so no missing values, lets convert to int and rename to more menaingful

In [1319]:
df = convert_to_int(df, 'WHEELS_OFF')

All non-missing  WHEELS_OFF  values are integers? {is_every_int}
converted successfully


In [1320]:
show_unique_missing_dtype(df, 'WHEELS_OFF')

WHEELS_OFF
-1     282493
707      4090
708      4046
706      4010
709      3960
        ...  
442         4
440         3
447         3
336         3
347         3
Name: count, Length: 1440, dtype: Int64

missing values:  0

datatype:  Int64


In [1321]:
df.rename(columns={'WHEELS_OFF': 'DEP_TAKEOFF_TIME'}, inplace=True)

## WHEELS_ON

In [1322]:
show_unique_missing_dtype(df, 'WHEELS_ON')

WHEELS_ON
-1.0       282493
 1850.0      2972
 1848.0      2953
 1854.0      2908
 1845.0      2893
            ...  
 329.0         14
 309.0         12
 341.0         12
 328.0         10
 324.0          8
Name: count, Length: 1441, dtype: int64

missing values:  1061

datatype:  float64


ok so lets deal with missing first. lets check if they are cancelled flights

In [1323]:
cancelled = check_if_cancelled(df, 'WHEELS_ON')

CANCELLED
0    628
1    433
Name: count, dtype: int64


In [1324]:
cancelled = cancelled[cancelled['CANCELLED'] == 1]
cancelled.isna().sum()

MONTH                        0
DAY_OF_MONTH                 0
DAY_OF_WEEK                  0
day_of_week_c                0
FLIGHT_DATE                  0
MARKETING_CARRIER_CODE       0
MARKETING_FLIGHT_NUMBER      0
AIRCRAFT_TAIL_NUMBER         0
ORIGIN_AIRPORT_CODE          0
ORIGIN_CITY_NAME             0
ORIGIN_STATE_CODE            0
ORIGIN_STATE_NAME            0
DEST_AIRPORT_CODE            0
DEST_CITY_NAME               0
DEST_STATE_CODE              0
DEST_STATE_NAME              0
SCHEDULED_DEP_TIME           0
SCHEDULED_DEP_HOUR           0
ACTUAL_DEP_TIME              0
ACTUAL_DEP_HOUR              0
DEP_DELAY                    0
DEP_DELAY_15_MIN             0
DEP_DELAY_GROUP              0
ORIGIN_TAXI_TIME             0
DEP_TAKEOFF_TIME             0
WHEELS_ON                  433
TAXI_IN                    433
CRS_ARR_TIME                 0
ARR_TIME                   433
ARR_DELAY                  433
ARR_DELAY_NEW              433
ARR_DEL15                  433
ARR_DELA

so this is new. first lets make the cancelled ones -1 to signify their time is not valid.

In [1325]:
cols_to_fill = ['WHEELS_ON', 'TAXI_IN', 'ARR_TIME', 'CARRIER_DELAY', 'WEATHER_DELAY', 'NAS_DELAY', 'SECURITY_DELAY', 'LATE_AIRCRAFT_DELAY']

In [1326]:
for col in cols_to_fill:
    missing_idx = cancelled.index
    df.loc[missing_idx, col] = -1

In [1327]:
cols_to_fill = ['ACTUAL_ELAPSED_TIME', 'AIR_TIME']

In [1328]:
for col in cols_to_fill:
    missing_idx = cancelled.index
    df.loc[missing_idx, col] = 0

In [1329]:
cancelled = check_if_cancelled(df, 'WHEELS_ON')
cancelled = cancelled[cancelled['CANCELLED'] == 1]
cancelled.isna().sum()

CANCELLED
0    628
Name: count, dtype: int64


MONTH                      0
DAY_OF_MONTH               0
DAY_OF_WEEK                0
day_of_week_c              0
FLIGHT_DATE                0
MARKETING_CARRIER_CODE     0
MARKETING_FLIGHT_NUMBER    0
AIRCRAFT_TAIL_NUMBER       0
ORIGIN_AIRPORT_CODE        0
ORIGIN_CITY_NAME           0
ORIGIN_STATE_CODE          0
ORIGIN_STATE_NAME          0
DEST_AIRPORT_CODE          0
DEST_CITY_NAME             0
DEST_STATE_CODE            0
DEST_STATE_NAME            0
SCHEDULED_DEP_TIME         0
SCHEDULED_DEP_HOUR         0
ACTUAL_DEP_TIME            0
ACTUAL_DEP_HOUR            0
DEP_DELAY                  0
DEP_DELAY_15_MIN           0
DEP_DELAY_GROUP            0
ORIGIN_TAXI_TIME           0
DEP_TAKEOFF_TIME           0
WHEELS_ON                  0
TAXI_IN                    0
CRS_ARR_TIME               0
ARR_TIME                   0
ARR_DELAY                  0
ARR_DELAY_NEW              0
ARR_DEL15                  0
ARR_DELAY_GROUP            0
ARR_TIME_BLK               0
CANCELLED     

now lets extract the remaining columns and just look at them first

In [1330]:
missing_rows = df[df['WHEELS_ON'].isna()]
missing_rows.isna().sum()

MONTH                        0
DAY_OF_MONTH                 0
DAY_OF_WEEK                  0
day_of_week_c                0
FLIGHT_DATE                  0
MARKETING_CARRIER_CODE       0
MARKETING_FLIGHT_NUMBER      0
AIRCRAFT_TAIL_NUMBER         0
ORIGIN_AIRPORT_CODE          0
ORIGIN_CITY_NAME             0
ORIGIN_STATE_CODE            0
ORIGIN_STATE_NAME            0
DEST_AIRPORT_CODE            0
DEST_CITY_NAME               0
DEST_STATE_CODE              0
DEST_STATE_NAME              0
SCHEDULED_DEP_TIME           0
SCHEDULED_DEP_HOUR           0
ACTUAL_DEP_TIME              0
ACTUAL_DEP_HOUR              0
DEP_DELAY                    0
DEP_DELAY_15_MIN             0
DEP_DELAY_GROUP              0
ORIGIN_TAXI_TIME             0
DEP_TAKEOFF_TIME             0
WHEELS_ON                  628
TAXI_IN                    628
CRS_ARR_TIME                 0
ARR_TIME                   628
ARR_DELAY                  628
ARR_DELAY_NEW              628
ARR_DEL15                  628
ARR_DELA

In [1331]:
missing_rows

MONTH  DAY_OF_MONTH  DAY_OF_WEEK day_of_week_c FLIGHT_DATE  \
14640        1            22            3           Wed  2020-01-22   
18541        1            13            1           Mon  2020-01-13   
32487        1            13            1           Mon  2020-01-13   
32544        1             8            3           Wed  2020-01-08   
32550        1            14            2           Tue  2020-01-14   
...        ...           ...          ...           ...         ...   
2667708      6            19            5           Fri  2020-06-19   
2667981      6            19            5           Fri  2020-06-19   
2673901      6            10            3           Wed  2020-06-10   
2704991      6             7            7           Sun  2020-06-07   
2743929      6            22            1           Mon  2020-06-22   

        MARKETING_CARRIER_CODE  MARKETING_FLIGHT_NUMBER AIRCRAFT_TAIL_NUMBER  \
14640                       AA                     4017               N843AE   
18541                       AA                     4150               N804AE   
32487                       HA                      628               N805HC   
32544                       HA                      638               N804HC   
32550                       HA                      638               N801HC   
...                        ...                      ...                  ...   
2667708                     WN                      969               N473WN   
2667981                     WN                     6865               N8673F   
2673901                     WN                      490               N8655D   
2704991                     AA                     4830               N629AE   
2743929                     WN                     1198               N7864B   

        ORIGIN_AIRPORT_CODE       ORIGIN_CITY_NAME ORIGIN_STATE_CODE  \
14640                   DFW  Dallas/Fort Worth, TX                TX   
18541                   DFW  Dallas/Fort Worth, TX                TX   
32487                   HNL           Honolulu, HI                HI   
32544                   HNL           Honolulu, HI                HI   
32550                   HNL           Honolulu, HI                HI   
...                     ...                    ...               ...   
2667708                 SAN          San Diego, CA                CA   
2667981                 STL          St. Louis, MO                MO   
2673901                 MCO            Orlando, FL                FL   
2704991                 CLT          Charlotte, NC                NC   
2743929                 BNA          Nashville, TN                TN   

        ORIGIN_STATE_NAME DEST_AIRPORT_CODE        DEST_CITY_NAME  \
14640               Texas               LAW  Lawton/Fort Sill, OK   
18541               Texas               LFT         Lafayette, LA   
32487              Hawaii               JHM           Kapalua, HI   
32544              Hawaii               JHM           Kapalua, HI   
32550              Hawaii               JHM           Kapalua, HI   
...                   ...               ...                   ...   
2667708        California               DAL            Dallas, TX   
2667981          Missouri               DAL            Dallas, TX   
2673901           Florida               SAT       San Antonio, TX   
2704991    North Carolina               MOB            Mobile, AL   
2743929         Tennessee               MCO           Orlando, FL   

        DEST_STATE_CODE DEST_STATE_NAME  SCHEDULED_DEP_TIME  \
14640                OK        Oklahoma                1035   
18541                LA       Louisiana                1229   
32487                HI          Hawaii                 806   
32544                HI          Hawaii                1013   
32550                HI          Hawaii                1013   
...                 ...             ...                 ...   
2667708              TX           Texas                1645 

so it seems like these flights left the origin but did not reach the destination, nor were they cancelled. looks like they were possibly diverted. thus we fill them with -1 because they do not exist. 

In [1332]:
cols_to_fill = ['WHEELS_ON', 'TAXI_IN', 'ARR_TIME']

In [1333]:
for col in cols_to_fill:
    missing_idx = missing_rows.index
    df.loc[missing_idx, col] = -1

In [1334]:
cols_to_fill = ['ACTUAL_ELAPSED_TIME', 'AIR_TIME']

In [1335]:
for col in cols_to_fill:
    missing_idx = missing_rows.index
    df.loc[missing_idx, col] = 0

recheck

In [1336]:
missing_rows = df[df['WHEELS_ON'].isna()]
missing_rows.shape

(0, 46)

convert to int

In [1337]:
df = convert_to_int(df, 'WHEELS_ON')

All non-missing  WHEELS_ON  values are integers? {is_every_int}
converted successfully


In [1338]:
show_unique_missing_dtype(df, 'WHEELS_ON')

WHEELS_ON
-1      283554
1850      2972
1848      2953
1854      2908
1845      2893
         ...  
329         14
309         12
341         12
328         10
324          8
Name: count, Length: 1441, dtype: Int64

missing values:  0

datatype:  Int64


rename it to something meaningful like `ARR_LANDING_TIME`

In [1339]:
df.rename(columns={'WHEELS_ON': 'ARR_LANDING_TIME'}, inplace=True)

## TAXI_IN

In [1340]:
show_unique_missing_dtype(df, 'TAXI_IN')

TAXI_IN
 4.0      415156
 5.0      362841
 3.0      324351
-1.0      283554
 6.0      282386
           ...  
 136.0         1
 112.0         1
 108.0         1
 111.0         1
 138.0         1
Name: count, Length: 136, dtype: int64

missing values:  0

datatype:  float64


rename to something similar as `TAXI_IN` to `DEST_TAXI_TIME` and convert to int

In [1341]:
df = convert_to_int(df, 'TAXI_IN')

All non-missing  TAXI_IN  values are integers? {is_every_int}
converted successfully


In [1342]:
df.rename(columns={'TAXI_IN': 'DEST_TAXI_TIME'}, inplace=True)

## CRS_ARR_TIME

In [1343]:
show_unique_missing_dtype(df, 'CRS_ARR_TIME')

CRS_ARR_TIME
1900    9526
1710    8655
1915    7491
1940    7277
1000    7273
        ... 
137        1
138        1
223        1
201        1
315        1
Name: count, Length: 1393, dtype: int64

missing values:  0

datatype:  int64


so same logic as `CRS_dEP_TIME`, lets extract its hour

In [1344]:
df = extract_hour(df, 'CRS_ARR_TIME', 'CRS_ARR_HOUR')

In [1345]:
show_unique_missing_dtype(df, 'CRS_ARR_HOUR')

CRS_ARR_HOUR
16    180483
19    179760
17    175672
10    172069
14    171397
11    170745
21    168110
12    167483
18    165390
13    164421
15    163198
9     153718
20    150308
8     142650
22    119987
23    113722
7      91247
0      35973
6      34085
5      14353
1       6455
4       2823
2       1233
3        552
24        13
Name: count, dtype: int64

missing values:  0

datatype:  int32


so again we see some 24 values. lets extract those and convert to 0

In [1346]:
# ensure the column is numeric
df['ARR_TIME_CLEAN'] = pd.to_numeric(df['CRS_ARR_TIME'], errors='coerce')

# filter rows with values = 2400
bad_times = df[df['ARR_TIME_CLEAN'] == 2400]
bad_times.shape

(13, 48)

In [1347]:
# ensure numeric dtype
df['CRS_ARR_TIME'] = pd.to_numeric(df['CRS_ARR_TIME'], errors='coerce').astype('Int64')

# replace exactly 2400 → 0
df.loc[df['CRS_ARR_TIME'] == 2400, 'CRS_ARR_TIME'] = 0

In [1348]:
# RECHECK

# ensure the column is numeric
df['ARR_TIME_CLEAN'] = pd.to_numeric(df['CRS_ARR_TIME'], errors='coerce')

# filter rows with values = 2400
bad_times = df[df['ARR_TIME_CLEAN'] == 2400]
bad_times.shape

(0, 48)

In [1349]:
# remove arr_time_clean

df = df.drop('ARR_TIME_CLEAN', axis=1)

In [1350]:
df = extract_hour(df, 'CRS_ARR_TIME', 'CRS_ARR_HOUR')

In [1351]:
show_unique_missing_dtype(df, 'CRS_ARR_HOUR')

CRS_ARR_HOUR
16    180483
19    179760
17    175672
10    172069
14    171397
11    170745
21    168110
12    167483
18    165390
13    164421
15    163198
9     153718
20    150308
8     142650
22    119987
23    113722
7      91247
0      35986
6      34085
5      14353
1       6455
4       2823
2       1233
3        552
Name: count, dtype: int64

missing values:  0

datatype:  int32


finally, lets rename the columns

In [1352]:
df.rename(columns={
    'CRS_ARR_TIME': 'SCHEDULED_ARR_TIME',
    'CRS_ARR_HOUR': 'SCHEDULED_ARR_HOUR'
}, inplace=True)

## ARR_TIME

In [1353]:
show_unique_missing_dtype(df, 'ARR_TIME')

ARR_TIME
-1.0       283554
 1857.0      2920
 1620.0      2915
 1855.0      2893
 1615.0      2888
            ...  
 342.0         12
 345.0         11
 339.0         11
 320.0         11
 324.0          7
Name: count, Length: 1441, dtype: int64

missing values:  0

datatype:  float64


seems fine. lets convert to int and also extract hours

In [1354]:
df = convert_to_int(df, 'ARR_TIME')

All non-missing  ARR_TIME  values are integers? {is_every_int}
converted successfully


In [1355]:
df = extract_hour(df, 'ARR_TIME', 'ARR_HOUR')

In [1356]:
show_unique_missing_dtype(df, 'ARR_HOUR')

ARR_HOUR
-1     283554
 16    163278
 18    156070
 11    152259
 14    150952
 10    150441
 19    150311
 17    149298
 15    148989
 12    148602
 13    147642
 20    142879
 21    142295
 9     140213
 8     127773
 22    108165
 23     91546
 7      88436
 6      36360
 0      33743
 5      14052
 1       8474
 4       5973
 2       2466
 3       1090
 24       986
Name: count, dtype: int64

missing values:  0

datatype:  int32


again we have 2400 values. lets repeat what we did in `CRS_ARR_TIME`

In [1357]:
# ensure the column is numeric
df['ARR_TIME_CLEAN'] = pd.to_numeric(df['ARR_TIME'], errors='coerce')

# filter rows with values = 2400
bad_times = df[df['ARR_TIME_CLEAN'] == 2400]
bad_times.shape

(986, 49)

In [1358]:
# ensure numeric dtype
df['ARR_TIME'] = pd.to_numeric(df['ARR_TIME'], errors='coerce').astype('Int64')

# replace exactly 2400 → 0
df.loc[df['ARR_TIME'] == 2400, 'ARR_TIME'] = 0

In [1359]:
# RECHECK

# ensure the column is numeric
df['ARR_TIME_CLEAN'] = pd.to_numeric(df['ARR_TIME'], errors='coerce')

# filter rows with values = 2400
bad_times = df[df['ARR_TIME_CLEAN'] == 2400]
bad_times.shape

(0, 49)

In [1360]:
# remove arr_time_clean

df = df.drop('ARR_TIME_CLEAN', axis=1)

In [1361]:
df = extract_hour(df, 'ARR_TIME', 'ARR_HOUR')

In [1362]:
show_unique_missing_dtype(df, 'ARR_HOUR')

ARR_HOUR
-1     283554
 16    163278
 18    156070
 11    152259
 14    150952
 10    150441
 19    150311
 17    149298
 15    148989
 12    148602
 13    147642
 20    142879
 21    142295
 9     140213
 8     127773
 22    108165
 23     91546
 7      88436
 6      36360
 0      34729
 5      14052
 1       8474
 4       5973
 2       2466
 3       1090
Name: count, dtype: int64

missing values:  0

datatype:  int32


In [1363]:
df.rename(columns={
    'ARR_TIME': 'ACTUAL_ARR_TIME',
    'ARR_HOUR': 'ACTUAL_ARR_HOUR'
}, inplace=True)

## ARR_DELAY

## ARR_DELAY_NEW

In [ ]:
show_unique_missing_dtype(df, 'ARR_DELAY_NEW')

again same reasoning as `dep_delay_new`, this column does not provide us new menaingful information as it is just repitition and can be achieved using a negative filter. better to drop. 

In [ ]:
df = df.drop('ARR_DELAY_NEW', axis=1)

## ARR_DEL15

In [ ]:
show_unique_missing_dtype(df, 'ARR_DEL15')

## ARR_DELAY_GROUP

## ARR_TIME_BLK

## CANCELLED

## CANCELLATION_CODE

In [ ]:
show_unique_missing_dtype(df, 'CANCELLATION_CODE')

so in our BI dashboard, a b c d e isnt very meaningful and readable. so here we see that this is not interpretable. we searched up and dug up the internet to get to know that from the U.S. Department of Transportation (DOT) flight data standards, here is what each cancellation code means:
- A = Carrier — airline-related (crew, aircraft, etc.)
- B = Weather — adverse weather at origin/destination or en route
- C = NAS — National Airspace System (ATC, delays, airport congestion)
- D = Security — e.g., TSA delays, threats, evacuations     

so instead of leaving it just like this, lets make it more visualizeable.

along with mapping each one, lets replace all null values with "Not Cancelled"

In [ ]:
cancel_map = {
    'A': 'Carrier',
    'B': 'Weather',
    'C': 'Airspace System (NAS)',
    'D': 'Security',
    'Not Cancelled': 'Not Cancelled'
}

df['CANCELLATION_CODE'] = df['CANCELLATION_CODE'].fillna('Not Cancelled').map(cancel_map)
df = to_string(df, 'CANCELLATION_CODE')

## CRS_ELAPSED_TIME

## ACTUAL_ELAPSED_TIME

## AIR_TIME

## DISTANCE

## DISTANCE_GROUP

## CARRIER_DELAY

In [ ]:
show_unique_missing_dtype(df, 'CARRIER_DELAY')

so this column is Carrier Delay (in Minutes). if it is null, then that probably means there was no `CARRIER_DELAY`. so we can just fill with 0. so first lets fill all cancelled with -1 and filled with 0. 

In [ ]:
def fill_delays(df, col):
    # fill all cancelled with -1
    cancelled = df[df['CANCELLED'] == 1] 
    print(cancelled[col].value_counts())
    print("\nmissing", cancelled[col].isna().sum(), "\n")
    missing_idx = cancelled.index
    df.loc[missing_idx, col] = -1
    
    # fill rest missing with 0
    missing_rows = df[df[col].isna()]
    missing_idx = missing_rows.index
    df.loc[missing_idx, col] = 0
    
    return df

In [ ]:
df = fill_delays(df, 'CARRIER_DELAY')
show_unique_missing_dtype(df, 'CARRIER_DELAY')

convert to int

In [ ]:
df = convert_to_int(df, 'CARRIER_DELAY')

In [ ]:
show_unique_missing_dtype(df, 'CARRIER_DELAY')

## WEATHER_DELAY

In [ ]:
show_unique_missing_dtype(df, 'WEATHER_DELAY')

so same as `CARRIER_DELAY`

In [ ]:
df = fill_delays(df, 'WEATHER_DELAY')
show_unique_missing_dtype(df, 'WEATHER_DELAY')

In [ ]:
df = convert_to_int(df, 'WEATHER_DELAY')
show_unique_missing_dtype(df, 'WEATHER_DELAY')

## NAS_DELAY

In [ ]:
show_unique_missing_dtype(df, 'NAS_DELAY')

In [ ]:
df = fill_delays(df, 'NAS_DELAY')
show_unique_missing_dtype(df, 'NAS_DELAY')

In [ ]:
df = convert_to_int(df, 'NAS_DELAY')
show_unique_missing_dtype(df, 'NAS_DELAY')

## SECURITY_DELAY

In [ ]:
show_unique_missing_dtype(df, 'SECURITY_DELAY')

In [ ]:
df = fill_delays(df, 'SECURITY_DELAY')
show_unique_missing_dtype(df, 'SECURITY_DELAY')

In [ ]:
df = convert_to_int(df, 'SECURITY_DELAY')
show_unique_missing_dtype(df, 'SECURITY_DELAY')

## LATE_AIRCRAFT_DELAY

In [ ]:
show_unique_missing_dtype(df, 'LATE_AIRCRAFT_DELAY')

In [ ]:
df = fill_delays(df, 'LATE_AIRCRAFT_DELAY')
show_unique_missing_dtype(df, 'LATE_AIRCRAFT_DELAY')

In [ ]:
df = convert_to_int(df, 'LATE_AIRCRAFT_DELAY')
show_unique_missing_dtype(df, 'LATE_AIRCRAFT_DELAY')

# Recheck
a simple recheck of all data statistics to see if the data is fine before donwloading

In [ ]:
df.shape

In [ ]:
# Check for missing values
df.isnull().sum()

In [ ]:
# Display data types
df.dtypes

In [ ]:
df

# Download cleaned dataset

In [ ]:
# Save the cleaned DataFrame to a CSV file
# df.to_csv('../data/flights_jantojun2020_3M_cleaned.csv', index=False)

In [ ]:
# Extract first 1000 rows
df_1000 = df.head(1000)

# Save to CSV with column headers
df_1000.to_csv('../data/flights_jantojun2020_1K_cleaned.csv', index=False)

In [ ]:
# revert pandas settings back to normal
# pd.reset_option('display.max_columns')

# Last run

In [ ]:
get_current_datetime()

In [ ]:
# Capture end time
end_time = datetime.now()

# Compute time difference
diff = end_time - start_time

# Total seconds (float)
total_seconds = diff.total_seconds()

# Decompose
hours, rem = divmod(total_seconds, 3600)
minutes, rem = divmod(rem, 60)
seconds = int(rem)
milliseconds = diff.microseconds // 1000

# Display
print(f"Start time : {start_time.strftime('%Y-%m-%d %H:%M:%S.%f')}")
print(f"End time   : {end_time.strftime('%Y-%m-%d %H:%M:%S.%f')}")
print(f"Duration   : {int(hours)}h {int(minutes)}m {seconds}s {milliseconds}ms")